---
title: "cibersortx_compare_DE"
author: "Melanie Smith"
date: "1 March 2024"
output: html_document
---

# Load required libraries

```{r load Libraries}
library(plyr)
library(reshape2)
library(dplyr)
library(tidyr)
library(stringr)
library(edgeR)
library(readxl)
library(readr)
library(magrittr)
library(tibble)
library(ggplot2)
library(ggbeeswarm)
library(ggrepel)
library(pheatmap)

# set project directory
projectDir <- "/home/smit1924/preeclampsia_sex_chromosome_informed"
# projectDir <- "/Users/pillmaka/Desktop/Melanie/drive-download"

expt_name <- 'cibersortx_de'


counts_geneID_file <- file.path(projectDir, "clearBoxRawData/rawCounts_melanie_v2.0.3.csv")
dge_list_obj_file <- file.path(projectDir, "noCovariate_output/DGEList_filtered_normalised.rds")
metadata_quality_file <- file.path(projectDir, "clearBoxCleanData/metadata_quality.csv")
allTable_female_file <- file.path(projectDir, "noCovariate_output/allTable_female.csv")
allTable_male_file <- file.path(projectDir, "noCovariate_output/allTable_male.csv")
topTable_male_file <- file.path(projectDir, "noCovariate_output/topTable_male_pe.csv")
gene_info_file <- file.path(projectDir, "clearBoxRawData/gene_id_vs_symbol.csv")

cibersortx_proportion_file <- file.path(projectDir, "cibersortx/outdir/CIBERSORTx_Adjusted.txt")
cibersortx_sigMatrix_file <- file.path(projectDir, "cibersortx/outdir/CIBERSORTx_sigmatrix_Adjusted.txt")

outdir <- file.path(projectDir, paste0(expt_name, "_output"))
outdir_literatureUp <- file.path(paste0(outdir, "/literatureUp"))
outdir_literatureDown <- file.path(paste0(outdir, "/literatureDown"))
outdir_deUp <- file.path(paste0(outdir, "/deUp"))
outdir_deDown <- file.path(paste0(outdir, "/deDown"))



dir.create(outdir)
dir.create(paste0(outdir, "/literatureUp"))
dir.create(paste0(outdir, "/literatureDown"))
dir.create(paste0(outdir, "/deUp"))
dir.create(paste0(outdir, "/deDown"))

# set max digits
options(digits=3)

`%notin%` <- Negate(`%in%`)

```

# Import all input data tables

```{r}

## DGEList object
dge_list_obj <- readRDS(dge_list_obj_file)
# create an lcpm object for later use
lcpm <- edgeR::cpm(dge_list_obj, log = TRUE)
metadata_quality <- read_csv(file = metadata_quality_file)

# DE tables (no covariate)
allTable_female <- read_csv(file = allTable_female_file)
allTable_male <- read_csv(file = allTable_male_file)
topTable_male <- read_csv(file = topTable_male_file)
# there are a couple of NAs in the gene symbols. I've looked up manually and will now replace the NA with the correct string
# Define replacements
replacements <- c(
  "ENSG00000279602" = "TEC",
  "ENSG00000260035" = "ENSG00000260035",
  "ENSG00000231205" = "ZNF826P",
  "ENSG00000265185" = "SNORD3B_1",
  "ENSG00000232316" = "LINC02518",
  "ENSG00000247809" = "NR2F2_AS1"
)
# Replace NA values with corresponding strings
# I looked each gene up on genecards
topTable_male$hgnc_symbol <- ifelse(is.na(topTable_male$hgnc_symbol), 
                                    replacements[topTable_male$ensembl_gene_id], 
                                    topTable_male$hgnc_symbol)

## Gene ID versus GeneSymbol
gene_info <- read_csv(file = gene_info_file)

## Gene ID versus GeneSymbol
gene_info <- read_csv(file = gene_info_file)
gene_info <- gene_info[!duplicated(gene_info$ensembl_gene_id), ]

# CIBERSORTx signature matrix and proportion files
cibersortx_proportion <- read.table(file = cibersortx_proportion_file, header = TRUE, sep = "\t") %>%
  dplyr::select(., samplename = Mixture, everything()) %>%
  dplyr::select(., -'P.value', -'Correlation', -'RMSE')
cibersortx_sigMatrix <- read.table(file = cibersortx_sigMatrix_file, header = TRUE, sep = "\t")
# remove rows that contain all zero counts in the signature matrix
cibersortx_sigMatrix <- cibersortx_sigMatrix[rowSums(cibersortx_sigMatrix[,-1])>0,]

```

# Cibersortx Cell-Type Proportions
## Plot cell-type proportions

```{r}
# which cell types are the most abundent
ordered_celltypes <- data.frame(proportion = 
  colSums(cibersortx_proportion[, -1])[order(colSums(cibersortx_proportion[, -1]),
                                             decreasing = TRUE)]) %>%
  # use the tibble package to rownames into the column names cell_type
  tibble::rownames_to_column("cell_type") 
  
# convert proportions table to long table format
data_long <- tidyr::pivot_longer(cibersortx_proportion,
                                 cols = colnames(cibersortx_proportion[, !colnames(cibersortx_proportion) %in% "samplename"]),
                                 names_to = "cell_type",
                                 values_to = "proportion")

# Generate 27 distinct colors from the viridis palette
my_palette <- colorspace::rainbow_hcl(27)

# Plot the stacked bar chart (not ordered)
ggplot(data_long, aes(x = samplename, y = proportion, fill = cell_type)) +
  geom_bar(stat = "identity") +
  labs(title = "CIBERSORTx Deconvolution", x = "Samplename", y = "Cell-Type Proportion") +
  theme_minimal() +
  theme(legend.position = "right") +
  viridis::scale_fill_viridis(discrete = TRUE, option = "C") +
  theme(axis.text.x = element_text(angle = 90, hjust = 1))

# Reorder samplenames based on the proportion of Fetal Cytotrophoblasts
ordered_samples_fetal_CTB <- dplyr::filter(data_long, cell_type == "Fetal.Cytotrophoblasts") %>%
  arrange(desc(proportion)) %>%
  dplyr::pull(samplename)

# Replot proprtions based on the proportion of Fetal B Cells
# png(filename=file.path(outdir, 'cibersortx_decon_CTB.png'), width=1200, height=800)
ggplot(data_long, aes(x = factor(samplename, levels = ordered_samples_fetal_CTB), y = proportion, fill = cell_type)) +
  geom_bar(stat = "identity") +
  labs(title = "CIBERSORTx Deconvolution (ordered by prop. CTB)",
       x = "Samplename",
       y = "Cell-Type Proportion") +
  theme_minimal() +
  theme(legend.position = "right") +
  viridis::scale_fill_viridis(discrete = TRUE, option = "C") +
  theme(axis.text.x = element_text(angle = 90, hjust = 1))
# graphics.off()

# Reorder samplenames based on the proportion of Maternal FCGR3A+ Monocytes
ordered_samples_maternal_mono <- dplyr::filter(data_long, cell_type == "Maternal.FCGR3A..Monocytes") %>%
  arrange(desc(proportion)) %>%
  dplyr::pull(samplename)

# Replot proprtions based on the proportion of Maternal FCGR3A+ Monocytes
# png(filename=file.path(outdir, 'cibersortx_decon_mat_mono.png'), width=1200, height=800)
ggplot(data_long, aes(x = factor(samplename, levels = ordered_samples_maternal_mono),
                      y = proportion,
                      fill = cell_type)) +
  geom_bar(stat = "identity") +
  labs(title = "CIBERSORTx Deconvolution (ordered by prop. Maternal FCGR3A+ Monocytes)", x = "Samplename", y = "Cell-Type Proportion") +
  theme_minimal() +
  theme(legend.position = "right") +
  viridis::scale_fill_viridis(discrete = TRUE, option = "C") +
  theme(axis.text.x = element_text(angle = 90, hjust = 1))
# graphics.off()

# Reorder samplenames based on the proportion of Maternal.Naive.CD8..T.Cells
ordered_samples_maternal_CD8_tcells <- dplyr::filter(data_long, cell_type == "Maternal.Naive.CD8..T.Cells") %>%
  arrange(desc(proportion)) %>%
  dplyr::pull(samplename)

# Replot proprtions based on the proportion of Maternal FCGR3A+ Monocytes
# png(filename=file.path(outdir, 'cibersortx_decon_mat_CD8_tcells.png'), width=1200, height=800)
ggplot(data_long, aes(x = factor(samplename, levels = ordered_samples_maternal_CD8_tcells),
                      y = proportion,
                      fill = cell_type)) +
  geom_bar(stat = "identity") +
  labs(title = "CIBERSORTx Deconvolution (ordered by prop. Maternal CD8+ T Cells)",
       x = "Samplename",
       y = "Cell-Type Proportion") +
  theme_minimal() +
  theme(legend.position = "right") +
  viridis::scale_fill_viridis(discrete = TRUE, option = "C") +
  theme(axis.text.x = element_text(angle = 90, hjust = 1))
# graphics.off()

# Heatmap of cell-type proportions
# add column annotation for maternal v fetal cells
# add row annotation for male v female samples
# add row annotation for PE v Uncomplicated

row_anno <- metadata_quality %>%
  dplyr::select(., samplename, Sex, Outcome, sex_outcome) %>%
  tibble::column_to_rownames("samplename")

column_anno <- data.frame(cell_type = colnames(cibersortx_sigMatrix[,-1]),
                          compartment = as.factor(colnames(cibersortx_sigMatrix[,-1]))) %>%
  tibble::column_to_rownames("cell_type")

# Split column names at the first period and extract the first element
compartment <- str_split_fixed(colnames(cibersortx_sigMatrix[, -1]), "\\.", n = 2)[, 1]

# Create the data frame
column_anno <- data.frame(
  cell_type = colnames(cibersortx_sigMatrix[, -1]),
  compartment = as.factor(compartment)
) %>%
  tibble::column_to_rownames("cell_type")

# plot the heatmap of all samples
# png(filename=file.path(outdir, 'cibersortx_heatmap_row_col_anno.png'), width=800, height=800)
pheatmap::pheatmap(cibersortx_proportion %>% tibble::column_to_rownames("samplename"),
                   annotation_row = row_anno,
                   annotation_col = column_anno,
                   scale = "row")
# graphics.off()

# plot the heatmap with SCP4060, SCP4010, SCP3492 removed
reduced_cibersortx_proportion <- subset(cibersortx_proportion, samplename %notin% c("SCP4060", "SCP4010", "SCP3492"))
rownames(reduced_cibersortx_proportion) <- NULL
png(filename=file.path(outdir, '58_sample_cibersortx_heatmap_row_col_anno.png'), width=800, height=800)
pheatmap::pheatmap(reduced_cibersortx_proportion %>%
                     tibble::column_to_rownames("samplename"),
                   annotation_row = row_anno,
                   annotation_col = column_anno,
                   scale = "row")
graphics.off()

```

# Sanity plot of genes
## PE associated genes from the literature

```{r}
# let's take a look at genes previously identified as associated with PE
# Gene of Interest (GOI)
# genes typically identified as up-regulated in the context of PE
GOI_PE_UP <- data.frame(ensbl=c("ENSG00000178726", "ENSG00000115602",
                             "ENSG00000174697", "ENSG00000070404",
                             "ENSG00000169495", "ENSG00000136002",
                             "ENSG00000012223", "ENSG00000142623",
                             "ENSG00000224658", "ENSG00000102755",
                             "ENSG00000118849", "ENSG00000165810",
                             "ENSG00000211448", "ENSG00000243836",
                             "ENSG00000150722", "ENSG00000181458",
                             "ENSG00000159399"),
                  gene=c("THBD", "IL1RL1", "LEP", "FSTL3",
                         "HTRA4", "ARHGEF4", "LTF", "PADI1",
                         "RP11-631F7.1", "FLT1", "RARRES1",
                         "BTNL9", "DIO2", "WDR86-AS1", "PPP1R1C",
                         "TMEM45A", "HK2"),
                  direction = "up")

# genes typically identified as down-regulated in the context of PE
GOI_PE_DOWN <- data.frame(ensbl=c("ENSG00000233913", "ENSG00000134548",
                                  "ENSG00000203859", "ENSG00000213088",
                                  "ENSG00000114473", "ENSG00000260719",
                                  "ENSG00000091137", "ENSG00000157064"),
                  gene=c("RPL10P9", "SPX", "HSD3B2", "ACKR1",
                         "IQCG", "AC009133.17", "SLC26A4", "NMNAT2"),
                  direction = "down")

samples <- dge_list_obj$samples
pd <- position_dodge(width = 0.5)

lcpm <- edgeR::cpm(dge_list_obj, log = TRUE)

## Print all of the up plots at once
for (i in 1:nrow(GOI_PE_UP)) {
melt_lcpm <- melt(subset(lcpm, rownames(lcpm) %in% GOI_PE_UP$ensbl[i])) %>% 
  as.data.frame() %>% 
  set_colnames(c("ensbl", "samplename", "log2cpm")) %>% 
  # tibble::rownames_to_column("samplename") %>% 
  dplyr::left_join(., dge_list_obj$samples[, c("samplename", "sex_outcome", "Sex")], by = "samplename") 

png(filename=file.path(outdir_literatureUp, paste(GOI_PE_UP$ensbl[i], GOI_PE_UP$gene[i], "png",sep = ".")),
    width=800,
    height=800)  
beeswarm <- ggplot(melt_lcpm, aes(x = Sex,
                                y = log2cpm,
                                fill = sex_outcome)) +
  stat_boxplot(geom="errorbar",
               position = pd,
               width = 0.2) +
  geom_boxplot(width = 0.5,
               position = pd) +
  stat_summary(fun = mean,
               geom = "point",
               size=2,
               shape = 5,
               position = position_dodge(0.5)) +
  geom_beeswarm(dodge.width = 0.6, cex = 2) +
  geom_text_repel(aes(label = samplename), size = 3, nudge_x = 0.1, nudge_y = 0.1) +
  theme_bw(base_size = 18) +
  ggtitle(paste(GOI_PE_UP$ensbl[i], GOI_PE_UP$gene[i], sep = " - ")) +
  ylab(paste("log2 CPM Counts")) +
  theme(axis.text.x = element_text(colour = "black")) +
  theme(legend.position = "bottom")
print(beeswarm)
graphics.off()
#print(beeswarm)
}

## Print all of the down plots at once
for (i in 1:nrow(GOI_PE_DOWN)) {
melt_lcpm <- melt(subset(lcpm, rownames(lcpm) %in% GOI_PE_DOWN$ensbl[i])) %>% 
  as.data.frame() %>% 
  set_colnames(c("ensbl", "samplename", "log2cpm")) %>% 
  # tibble::rownames_to_column("samplename") %>% 
  dplyr::left_join(., dge_list_obj$samples[, c("samplename", "sex_outcome", "Sex")], by = "samplename") 

png(filename=file.path(outdir_literatureDown, paste(GOI_PE_DOWN$ensbl[i], GOI_PE_DOWN$gene[i], "png",sep = ".")),
    width=800,
    height=800)  
beeswarm <- ggplot(melt_lcpm, aes(x = Sex,
                                y = log2cpm,
                                fill = sex_outcome)) +
  stat_boxplot(geom="errorbar",
               position = pd,
               width = 0.2) +
  geom_boxplot(width = 0.5,
               position = pd) +
  stat_summary(fun = mean,
               geom = "point",
               size=2,
               shape = 5,
               position = position_dodge(0.5)) +
  geom_beeswarm(dodge.width = 0.6, cex = 2) +
  geom_text_repel(aes(label = samplename), size = 3, nudge_x = 0.1, nudge_y = 0.1) +
  theme_bw(base_size = 18) +
  ggtitle(paste(GOI_PE_DOWN$ensbl[i], GOI_PE_DOWN$gene[i], sep = " - ")) +
  ylab(paste("log2 CPM Counts")) +
  theme(axis.text.x = element_text(colour = "black")) +
  theme(legend.position = "bottom")
print(beeswarm)
graphics.off()
#print(beeswarm)
}
```

## DE genes from my analysis (no covariate model)

```{r}
# let's take a look at the DE genes with the highest logFC that also have a "decent" expression level

GOI_PE_male_UP <- topTable_male %>%
  dplyr::filter(logFC > 0 & AveExpr > 3) %>%
  dplyr::select(ensbl = ensembl_gene_id, gene = hgnc_symbol) %>%
  dplyr::mutate(direction = "up")

GOI_PE_male_DOWN <- topTable_male %>%
  dplyr::filter(logFC < 0 & AveExpr > 3) %>%
  dplyr::select(ensbl = ensembl_gene_id, gene = hgnc_symbol) %>%
  dplyr::mutate(direction = "down")


samples <- dge_list_obj$samples
pd <- position_dodge(width = 0.5)

lcpm <- edgeR::cpm(dge_list_obj, log = TRUE)

## Print all of the DE Up plots at once
for (i in 1:nrow(GOI_PE_male_UP)) {
melt_lcpm <- reshape2::melt(subset(lcpm, rownames(lcpm) %in% GOI_PE_male_UP$ensbl[i])) %>% 
  as.data.frame() %>% 
  set_colnames(c("ensbl", "samplename", "log2cpm")) %>% 
  dplyr::left_join(., dge_list_obj$samples[, c("samplename", "Outcome", "Sex")], by = "samplename")  

png(filename=file.path(outdir_deUp, paste(GOI_PE_male_UP$ensbl[i], GOI_PE_male_UP$gene[i], "png",sep = ".")),
    width=800,
    height=800) 

beeswarm <- ggplot(melt_lcpm %>% dplyr::mutate(., sex_outcome = paste(Sex, Outcome, sep = "_")),
       aes(x = sex_outcome,
                      y = log2cpm,
                      label = samplename)) +
  stat_boxplot(geom = "errorbar",
               position = pd,
               width = 0.2) +
  geom_boxplot(width = 0.5,
               position = pd) +
  stat_summary(fun = mean,
               geom = "point",
               size=2,
               shape = 5,
               position = position_dodge(0.5)) +
  geom_point(position = position_jitter(width = 0.1),
             aes(x = sex_outcome,
                 y = log2cpm,
                 colour = Outcome)) +
  stat_summary(fun = mean,
               geom = "point",
               size=2,
               shape = 5,
               position = position_dodge(0.5)) +
  geom_text_repel(data = melt_lcpm %>% dplyr::mutate(., sex_outcome = paste(Sex, Outcome, sep = "_")),
                  aes(label = samplename,
                      colour = Outcome),
                  box.padding = unit(0.3, "lines"),
                  point.padding = unit(0.5, "lines"),
                  max.overlaps = Inf,
                  # fill = "white",
                  segment.color = "black",
                  segment.size = 0.5) +
  theme_bw(base_size = 18) +
  ggtitle(paste(GOI_PE_male_UP$ensbl[i], GOI_PE_male_UP$gene[i], sep = " - ")) +
  ylab(paste("log2 CPM")) +
  theme(axis.text.x = element_text(colour = "black")) +
  theme(legend.position = "bottom")
print(beeswarm)
graphics.off()
}

## Print all of the DE Down plots at once
for (i in 1:nrow(GOI_PE_male_DOWN)) {
melt_lcpm <- reshape2::melt(subset(lcpm, rownames(lcpm) %in% GOI_PE_male_DOWN$ensbl[i])) %>% 
  as.data.frame() %>% 
  set_colnames(c("ensbl", "samplename", "log2cpm")) %>% 
  dplyr::left_join(., dge_list_obj$samples[, c("samplename", "Outcome", "Sex")], by = "samplename")  

png(filename=file.path(outdir_deDown, paste(GOI_PE_male_DOWN$ensbl[i], GOI_PE_male_DOWN$gene[i], "png",sep = ".")),
    width=800,
    height=800) 

beeswarm <- ggplot(melt_lcpm %>% dplyr::mutate(., sex_outcome = paste(Sex, Outcome, sep = "_")),
       aes(x = sex_outcome,
                      y = log2cpm,
                      label = samplename)) +
  stat_boxplot(geom = "errorbar",
               position = pd,
               width = 0.2) +
  geom_boxplot(width = 0.5,
               position = pd) +
  stat_summary(fun = mean,
               geom = "point",
               size=2,
               shape = 5,
               position = position_dodge(0.5)) +
  geom_point(position = position_jitter(width = 0.1),
             aes(x = sex_outcome,
                 y = log2cpm,
                 colour = Outcome)) +
  stat_summary(fun = mean,
               geom = "point",
               size=2,
               shape = 5,
               position = position_dodge(0.5)) +
  geom_text_repel(data = melt_lcpm %>% dplyr::mutate(., sex_outcome = paste(Sex, Outcome, sep = "_")),
                  aes(label = samplename,
                      colour = Outcome),
                  box.padding = unit(0.3, "lines"),
                  point.padding = unit(0.5, "lines"),
                  max.overlaps = Inf,
                  # fill = "white",
                  segment.color = "black",
                  segment.size = 0.5) +
  theme_bw(base_size = 18) +
  ggtitle(paste(GOI_PE_male_DOWN$ensbl[i], GOI_PE_male_DOWN$gene[i], sep = " - ")) +
  ylab(paste("log2 CPM")) +
  theme(axis.text.x = element_text(colour = "black")) +
  theme(legend.position = "bottom")
print(beeswarm)
graphics.off()
}

```

## DE genes overlap with signature matrix

```{r}
# lets take a look at the DE topTable and see how many of the DE genes are part of the signature matrix
sig_de_overlap <- subset(cibersortx_sigMatrix, GeneSymbol %in% topTable_male$hgnc_symbol)

# lets take a look at IGFBP7 and VCAN
# in both genes SCP4157 is highly expressed and seems to be driving the DE
# also, in both genes STP0148 is more lowly expressed than the other PE boys.

melt_lcpm <- reshape2::melt(subset(lcpm, rownames(lcpm) %in% "ENSG00000038427")) %>% 
  as.data.frame() %>% 
  set_colnames(c("ensbl", "samplename", "log2cpm")) %>% 
  dplyr::left_join(., dge_list_obj$samples[, c("samplename", "Outcome", "Sex")], by = "samplename") %>%
  dplyr::left_join(., cibersortx_proportion[, c("samplename", "Fetal.Mesenchymal.Stem.Cells")], by = "samplename") %>%
  dplyr::mutate(sex_outcome = paste(Sex, Outcome, sep = "_"))

ggplot2::ggplot(data = melt_lcpm,
                aes(x = Fetal.Mesenchymal.Stem.Cells,
                    y = log2cpm,
                    colour = sex_outcome)) +
  geom_point()

# what is the gene signature for the gene of interest according to the cibersort signature matrix
signature_GOI_expression <- dplyr::filter(sig_de_overlap, GeneSymbol == "VCAN") %>%
  tibble::column_to_rownames("GeneSymbol") %>%
  t() %>%
  data.frame() %>%
  tibble::rownames_to_column("cell_type")

```

# Correlate gene expression with cell-type proportions
## Only genes which are both DE and in the cibersortx signature matrix

```{r}

# lcpm
lcpm <- edgeR::cpm(dge_list_obj, log = TRUE)
# overlap between signature and DE genes - this is used to cut down the number of genes tested for correlation
sig_de_overlap <- subset(cibersortx_sigMatrix, GeneSymbol %in% topTable_male$hgnc_symbol)

# Initialize an empty data frame to store correlation results
final_correlation_results <- data.frame(Cell_Type = names(sig_de_overlap)[-1], stringsAsFactors = FALSE)

# loop through the genes and correlate with cell-type proportions
for (i in 1:nrow(sig_de_overlap)) {
  
  sig_de_overlap_ensbl <- dplyr::left_join(sig_de_overlap, topTable_male[, c("ensembl_gene_id", "hgnc_symbol")], by = c("GeneSymbol" = "hgnc_symbol")) %>%
    dplyr::select(., GeneSymbol, ensembl_gene_id, everything())
  
  GOI_ensbl <- paste0(sig_de_overlap_ensbl$ensembl_gene_id[i])
  GOI_hgnc <- paste0(sig_de_overlap_ensbl$GeneSymbol[i])
  
  cibersortx_proportion_rownames <- reduced_cibersortx_proportion %>%
    tibble::column_to_rownames("samplename")
  
  # subset the GOI lcpm expression information for all samples
  lcpm_GOI <- reshape2::melt(subset(lcpm, rownames(lcpm) %in% sig_de_overlap_ensbl$ensembl_gene_id[i])) %>% 
    as.data.frame() %>% 
    set_colnames(c("ensbl", "samplename", paste0(sig_de_overlap_ensbl$ensembl_gene_id[i]))) %>% 
    dplyr::select(., -ensbl) %>%
    dplyr::left_join(., reduced_cibersortx_proportion, by = "samplename") %>%
    dplyr::select(., -samplename)
  
  # Calculate correlations between gene expression and cell type proportions for each cell type
  correlations <- lapply(colnames(lcpm_GOI)[-1], function(cell_type) {
    cor(lcpm_GOI[[1]], lcpm_GOI[[cell_type]])
  })
  
  # Combine correlations into a data frame
  correlation_results <- data.frame(names(lcpm_GOI)[-1])
  
  # Add correlation results for the current gene as new columns
  colnames(correlation_results) <- paste(GOI_hgnc)
  correlation_results[] <- unlist(correlations)
  
  # Append the current gene's correlation results to the final data frame
  final_correlation_results <- cbind(final_correlation_results, correlation_results)
  
}


```

## All genes which are DE in the male comparison

```{r}

# lcpm
lcpm <- edgeR::cpm(dge_list_obj, log = TRUE)
# overlap between signature and DE genes - this is used to cut down the number of genes tested for correlation
sig_de_overlap <- subset(cibersortx_sigMatrix, GeneSymbol %in% topTable_male$hgnc_symbol)

# Initialize an empty data frame to store correlation results
final_correlation_results_all <- data.frame(Cell_Type = names(cibersortx_sigMatrix)[-1], stringsAsFactors = FALSE)

# loop through the genes and correlate with cell-type proportions
for (i in 1:nrow(topTable_male)) {
  
  de_genes_symbol_ensbl <- topTable_male %>%
    dplyr::select(., GeneSymbol = hgnc_symbol, ensembl_gene_id)
  
  GOI_ensbl <- paste0(de_genes_symbol_ensbl$ensembl_gene_id[i])
  GOI_hgnc <- paste0(de_genes_symbol_ensbl$GeneSymbol[i])
  
  cibersortx_proportion_rownames <- reduced_cibersortx_proportion %>%
    tibble::column_to_rownames("samplename")
  
  # subset the GOI lcpm expression information for all samples
  lcpm_GOI <- reshape2::melt(subset(lcpm, rownames(lcpm) %in% de_genes_symbol_ensbl$ensembl_gene_id[i])) %>% 
    as.data.frame() %>% 
    set_colnames(c("ensbl", "samplename", paste0(de_genes_symbol_ensbl$ensembl_gene_id[i]))) %>% 
    dplyr::select(., -ensbl) %>%
    dplyr::left_join(., reduced_cibersortx_proportion, by = "samplename") %>%
    dplyr::select(., -samplename)
  
  # Calculate correlations between gene expression and cell type proportions for each cell type
  correlations_all <- lapply(colnames(lcpm_GOI)[-1], function(cell_type) {
    cor(lcpm_GOI[[1]], lcpm_GOI[[cell_type]])
  })
  
  # Combine correlations into a data frame
  correlation_results_all <- data.frame(names(lcpm_GOI)[-1])
  
  # Add correlation results for the current gene as new columns
  colnames(correlation_results_all) <- paste(GOI_hgnc)
  correlation_results_all[] <- unlist(correlations_all)
  
  # Append the current gene's correlation results to the final data frame
  final_correlation_results_all <- cbind(final_correlation_results_all, correlation_results_all)
  
}

t_final_correlation_results_all <- final_correlation_results_all %>%
  tibble::column_to_rownames("Cell_Type") %>%
  t()
```

# Heatmaps
## Heatmap of cell-type and DE gene expression

```{r}
# add column annotation so I can see if the gene is up or down regulated in the DE
column_annotation_overlap <- subset(topTable_male, hgnc_symbol %in% sig_de_overlap$GeneSymbol) %>%
  dplyr::mutate(DE = factor(ifelse(logFC < 0, "down",
                            "up"))) %>%
  tibble::column_to_rownames("hgnc_symbol") %>%
  dplyr::select(., DE)

# only DE genes that are in the gene signature
# png(file=file.path(outdir, "heatmap_signature_DE_cellType_exp_cor.png"),
#     width=12,
#     height=8,
#     units = "in",
#     res = 150)
pheatmap(final_correlation_results %>% tibble::column_to_rownames("Cell_Type"),
         treeheight_row = 0,
         treeheight_col = 0,
         cutree_rows = 3,
         cutree_cols = 2,
         annotation_col = column_annotation_overlap)
# graphics.off()

# add column annotation so I can see if the gene is up or down regulated in the DE
column_annotation <- topTable_male %>%
  dplyr::mutate(DE = factor(ifelse(logFC < 0, "down",
                            "up"))) %>%
  tibble::column_to_rownames("hgnc_symbol") %>%
  dplyr::select(., DE)

# all DE genes
# png(file=file.path(outdir, "heatmap_all_DE_cellType_exp_cor.png"),
#     width=22,
#     height=8,
#     units = "in",
#     res = 150)
pheatmap(final_correlation_results_all %>% tibble::column_to_rownames("Cell_Type"),
         treeheight_row = 0,
         treeheight_col = 0,
         cutree_rows = 3,
         cutree_cols = 2,
         annotation_col = column_annotation)
# graphics.off()


```

# Session information

```{r session info}
sessionInfo()
```
