---
title: "cibersortx_reduced_sigMatrix"
author: "Melanie Smith"
date: "27 March 2024"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Load required libraries

```{r load Libraries}
library(plyr)
library(reshape2)
library(dplyr)
library(tidyr)
library(stringr)
library(edgeR)
library(readxl)
library(readr)
library(magrittr)
library(tibble)
library(ggplot2)
library(ggbeeswarm)
library(ggrepel)
library(pheatmap)

# set project directory
projectDir <- "/home/smit1924/preeclampsia_sex_chromosome_informed"
# projectDir <- "/Users/pillmaka/Desktop/Melanie/drive-download"

# set experiment name
expt_name <- 'cibersortx_reduced_sigMatrix'

# create objects for files to be imported
counts_geneID_file <- file.path(projectDir, "clearBoxRawData/rawCounts_melanie_v2.0.3.csv")
dge_list_obj_file <- file.path(projectDir, "noCovariate_output/DGEList_filtered_normalised.rds")
tpm_table_file <- file.path(projectDir, "clearBoxCleanData/tpm_bulkCountMatrix_all_gene_symbol.txt")
metadata_quality_file <- file.path(projectDir, "clearBoxCleanData/metadata_quality.csv")
allTable_female_file <- file.path(projectDir, "noCovariate_output/allTable_female.csv")
allTable_male_file <- file.path(projectDir, "noCovariate_output/allTable_male.csv")
topTable_male_file <- file.path(projectDir, "noCovariate_output/topTable_male_pe.csv")
gene_info_file <- file.path(projectDir, "clearBoxRawData/gene_id_vs_symbol.csv")

cibersortx_proportion_file <- file.path(projectDir, "cibersortx/outdir/CIBERSORTx_Adjusted.txt")
cibersortx_sigMatrix_file <- file.path(projectDir, "cibersortx/outdir/CIBERSORTx_sigmatrix_Adjusted.txt")
cibersortx_reduced_proportion_file <- file.path(projectDir, "cibersortx_reduced_sigMatrix/outdir/CIBERSORTx_Adjusted.txt")
cibersortx_reduced_sigMatrix_file <- file.path(projectDir, "cibersortx_reduced_sigMatrix/outdir/CIBERSORTx_sigmatrix_Adjusted.txt")
cibersortx_reduced_mixture_adjusted_file <- file.path(projectDir, "cibersortx_reduced_sigMatrix/outdir/CIBERSORTx_Mixtures_Adjusted.txt")

# create objects for output files
outdir <- file.path(projectDir, paste0(expt_name, "_output"))
outdir_literatureUp <- file.path(paste0(outdir, "/literatureUp"))
outdir_literatureDown <- file.path(paste0(outdir, "/literatureDown"))
outdir_deUp <- file.path(paste0(outdir, "/deUp"))
outdir_deDown <- file.path(paste0(outdir, "/deDown"))



dir.create(outdir)
dir.create(paste0(outdir, "/literatureUp"))
dir.create(paste0(outdir, "/literatureDown"))
dir.create(paste0(outdir, "/deUp"))
dir.create(paste0(outdir, "/deDown"))

# set max digits
options(digits=3)

`%notin%` <- Negate(`%in%`)

```

# Import all input data tables

```{r}

## DGEList object
dge_list_obj <- readRDS(dge_list_obj_file)
# create a log 2 cpm object for later use
lcpm <- edgeR::cpm(dge_list_obj, log = TRUE)
# create an un-logged cpm object for later use
cpm <- edgeR::cpm(dge_list_obj, log = FALSE)
metadata_quality <- read_csv(file = metadata_quality_file)

## import the tpm table - all genes, all samples
tpm_bulkCountMatrix_all_genes <- read.table(file = tpm_table_file, header = TRUE, sep = "\t")
dim(tpm_bulkCountMatrix_all_genes) # 13784    62

# DE tables (no covariate)
allTable_female <- read_csv(file = allTable_female_file)
allTable_male <- read_csv(file = allTable_male_file)
topTable_male <- read_csv(file = topTable_male_file)
# there are a couple of NAs in the gene symbols. I've looked up manually and will now replace the NA with the correct string
# Define replacements
replacements <- c(
  "ENSG00000279602" = "TEC",
  "ENSG00000260035" = "ENSG00000260035",
  "ENSG00000231205" = "ZNF826P",
  "ENSG00000265185" = "SNORD3B_1",
  "ENSG00000232316" = "LINC02518",
  "ENSG00000247809" = "NR2F2_AS1"
)
# Replace NA values with corresponding strings
# I looked each gene up on genecards
topTable_male$hgnc_symbol <- ifelse(is.na(topTable_male$hgnc_symbol), 
                                    replacements[topTable_male$ensembl_gene_id], 
                                    topTable_male$hgnc_symbol)

## Gene ID versus GeneSymbol
gene_info <- read_csv(file = gene_info_file)

## Gene ID versus GeneSymbol
gene_info <- read_csv(file = gene_info_file)
gene_info <- gene_info[!duplicated(gene_info$ensembl_gene_id), ]

# CIBERSORTx signature matrix and proportion files - reduced cell types
cibersortx_reduced_proportion <- read.table(file = cibersortx_reduced_proportion_file, header = TRUE, sep = "\t") %>%
  dplyr::select(., samplename = Mixture, everything()) %>%
  dplyr::select(., -'P.value', -'Correlation', -'RMSE')

# CIBERSORTx proportion file with RMSE  - reduced cell types
cibersortx_reduced_proportion_rmse <- read.table(file = cibersortx_reduced_proportion_file, header = TRUE, sep = "\t") %>%
  dplyr::select(., samplename = Mixture, everything())

cibersortx_reduced_sigMatrix <- read.table(file = cibersortx_reduced_sigMatrix_file, header = TRUE, sep = "\t")
# remove rows that contain all zero counts in the signature matrix
cibersortx_reduced_sigMatrix <- cibersortx_reduced_sigMatrix[rowSums(cibersortx_reduced_sigMatrix[,-1])>0,]
dim(cibersortx_reduced_sigMatrix)

# CIBERSORTx adjusted mixture - reduced cell types
cibersortx_reduced_mixture_adjusted <- read.table(file = cibersortx_reduced_mixture_adjusted_file, header = TRUE, sep = "\t")

```

# Sanity plots
## Plot the Cibersortx Signature Matrix as a heatmap
### Reduced signature matrix

```{r}

# Cibersort gene signature
# png(file=file.path(outdir, "heatmap_cibersortx_reduced signature_matrix.png"),
#     width=12,
#     height=8,
#     units = "in",
#     res = 150)
cibersortx_reduced_sigMatrix %>%
  `rownames<-`(NULL) %>%
  as.data.frame() %>%
  tibble::column_to_rownames("GeneSymbol") %>%
  pheatmap(scale = "row",
           show_rownames = FALSE,
           treeheight_row = 0,
           main = "Cibersortx Reduced Adjusted Signature Matrix (scaled by row)")
# dev.off()
```

# Cibersortx Cell-Type Proportions
## Plot cell-type proportions

```{r}
# which cell types are the most abundent
ordered_celltypes <- data.frame(proportion = 
  colSums(cibersortx_reduced_proportion[, -1])[order(colSums(cibersortx_reduced_proportion[, -1]),
                                             decreasing = TRUE)]) %>%
  # use the tibble package to rownames into the column names cell_type
  tibble::rownames_to_column("cell_type") 
  
# convert proportions table to long table format
data_long <- tidyr::pivot_longer(cibersortx_reduced_proportion,
                                 cols = colnames(cibersortx_reduced_proportion[, !colnames(cibersortx_reduced_proportion) %in% "samplename"]),
                                 names_to = "cell_type",
                                 values_to = "proportion")

# Generate 27 distinct colors from the viridis palette
my_palette <- colorspace::rainbow_hcl(27)

# Plot the stacked bar chart (not ordered)
ggplot(data_long, aes(x = samplename, y = proportion, fill = cell_type)) +
  geom_bar(stat = "identity") +
  labs(title = "CIBERSORTx Deconvolution", x = "Samplename", y = "Cell-Type Proportion") +
  theme_minimal() +
  theme(legend.position = "right") +
  viridis::scale_fill_viridis(discrete = TRUE, option = "C") +
  theme(axis.text.x = element_text(angle = 90, hjust = 1))

# Reorder samplenames based on the proportion of Fetal Cytotrophoblasts
ordered_samples_fetal_CTB <- dplyr::filter(data_long, cell_type == "Fetal.Cytotrophoblasts") %>%
  arrange(desc(proportion)) %>%
  dplyr::pull(samplename)

# Replot proprtions based on the proportion of Fetal B Cells
# png(filename=file.path(outdir, 'cibersortx_decon_CTB.png'), width=1200, height=800)
ggplot(data_long, aes(x = factor(samplename, levels = ordered_samples_fetal_CTB), y = proportion, fill = cell_type)) +
  geom_bar(stat = "identity") +
  labs(title = "CIBERSORTx Deconvolution (ordered by prop. CTB)",
       x = "Samplename",
       y = "Cell-Type Proportion") +
  theme_minimal() +
  theme(legend.position = "right") +
  viridis::scale_fill_viridis(discrete = TRUE, option = "C") +
  theme(axis.text.x = element_text(angle = 90, hjust = 1))
# graphics.off()

# Reorder samplenames based on the proportion of Maternal FCGR3A+ Monocytes
ordered_samples_maternal_mono <- dplyr::filter(data_long, cell_type == "Maternal.FCGR3A..Monocytes") %>%
  arrange(desc(proportion)) %>%
  dplyr::pull(samplename)

# Replot proprtions based on the proportion of Maternal FCGR3A+ Monocytes
# png(filename=file.path(outdir, 'cibersortx_decon_mat_mono.png'), width=1200, height=800)
ggplot(data_long, aes(x = factor(samplename, levels = ordered_samples_maternal_mono),
                      y = proportion,
                      fill = cell_type)) +
  geom_bar(stat = "identity") +
  labs(title = "CIBERSORTx Deconvolution (ordered by prop. Maternal FCGR3A+ Monocytes)", x = "Samplename", y = "Cell-Type Proportion") +
  theme_minimal() +
  theme(legend.position = "right") +
  viridis::scale_fill_viridis(discrete = TRUE, option = "C") +
  theme(axis.text.x = element_text(angle = 90, hjust = 1))
# graphics.off()

# Reorder samplenames based on the proportion of Maternal.Naive.CD8..T.Cells
ordered_samples_CD8_tcells <- dplyr::filter(data_long, cell_type == "Naive.CD8..T.Cells") %>%
  arrange(desc(proportion)) %>%
  dplyr::pull(samplename)

# Replot proprtions based on the proportion of Maternal FCGR3A+ Monocytes
# png(filename=file.path(outdir, 'cibersortx_decon_mat_CD8_tcells.png'), width=1200, height=800)
ggplot(data_long, aes(x = factor(samplename, levels = ordered_samples_CD8_tcells),
                      y = proportion,
                      fill = cell_type)) +
  geom_bar(stat = "identity") +
  labs(title = "CIBERSORTx Deconvolution (ordered by prop. CD8+ T Cells)",
       x = "Samplename",
       y = "Cell-Type Proportion") +
  theme_minimal() +
  theme(legend.position = "right") +
  viridis::scale_fill_viridis(discrete = TRUE, option = "C") +
  theme(axis.text.x = element_text(angle = 90, hjust = 1))
# graphics.off()

# Heatmap of cell-type proportions
# add column annotation for maternal v fetal cells
# add row annotation for male v female samples
# add row annotation for PE v Uncomplicated

row_anno <- metadata_quality %>%
  dplyr::select(., samplename, Sex, Outcome, sex_outcome) %>%
  dplyr::left_join(., cibersortx_reduced_proportion_rmse %>% dplyr::select(., samplename, Correlation, RMSE), by = "samplename") %>%
  tibble::column_to_rownames("samplename")

# plot the heatmap of all samples
# png(filename=file.path(outdir, 'cibersortx_heatmap_row_col_anno.png'), width=800, height=800)
pheatmap::pheatmap(cibersortx_reduced_proportion %>% tibble::column_to_rownames("samplename"),
                   annotation_row = row_anno,
                   # annotation_col = column_anno,
                   scale = "row")
# graphics.off()

# plot the heatmap with SCP4060, SCP4010, SCP3492 removed
cibersortx_reduced_proportion_reduced_samples <- subset(cibersortx_reduced_proportion, samplename %notin% c("SCP4060", "SCP4010", "SCP3492"))
rownames(cibersortx_reduced_proportion_reduced_samples) <- NULL

# save a version of the heatmap with raw proportion data as the input
# png(filename=file.path(outdir, '58_sample_cibersortx_heatmap_row_col_anno_raw.png'), width=800, height=800)
pheatmap::pheatmap(cibersortx_reduced_proportion_reduced_samples %>%
                     tibble::column_to_rownames("samplename"),
                   annotation_row = row_anno,
                   color = colorRampPalette(c("white", "red"))(27),
                   main = "raw proportion data")
# graphics.off()
# save a version of the heatmap with logged proportion data as the input
# save which columns are numeric
id_numeric <- sapply(cibersortx_reduced_proportion_reduced_samples, is.numeric)
# make a copy of the raw data to transform
log2_cibersortx_reduced_proportion_reduced_samples <- cibersortx_reduced_proportion_reduced_samples
# transform the raw data
log2_cibersortx_reduced_proportion_reduced_samples[, id_numeric] <- log2(cibersortx_reduced_proportion_reduced_samples[, id_numeric] + .01)
# save the plot
# png(filename=file.path(outdir, '58_sample_cibersortx_heatmap_row_col_anno_logged.png'), width=800, height=800)
pheatmap::pheatmap(log2_cibersortx_reduced_proportion_reduced_samples %>%
                     tibble::column_to_rownames("samplename"),
                   annotation_row = row_anno,
                   color = colorRampPalette(c("white", "blue"))(27),
                   main = "Log2 (+.01) proportion data")
# graphics.off()

# save a version of the heatmap with mean centred and scaled proportions
# create a new matrix for the scaled data
cibersortx_reduced_proportion_reduced_samples_scaled <- cibersortx_reduced_proportion_reduced_samples %>%
  tibble::column_to_rownames("samplename") %>%
  t() %>%
  scale(., center = TRUE, scale = TRUE) %>%
  t()

# png(filename=file.path(outdir, '58_sample_cibersortx_heatmap_row_col_anno_scaled.png'), width=800, height=800)
pheatmap::pheatmap(cibersortx_reduced_proportion_reduced_samples_scaled,
                   annotation_row = row_anno,
                   main = "mean centred and scaled proportion data")
# graphics.off()

```

# Predicted gene expression
## Scatter plot of the adjusted matrix versus the sample tpms

```{r}
# there are 26 genes in the tpm matrix with duplicated gene names as an artifact of the original join - delete one and move on with life initial dim 13784 x 62
# Find the indices of the first occurrence of each unique gene name
first_indices <- !duplicated(tpm_bulkCountMatrix_all_genes$hgnc_symbol)

# Subset the matrix to keep only the rows with the first occurrence of duplicated gene names
tpm <- tpm_bulkCountMatrix_all_genes[first_indices, ]
dim(tpm) # 13764 62
# remove the samples not to be considered here
tpm <- tpm %>%
  dplyr::select(., -SCP4060, -SCP4010, -SCP3492) %>%
  # rename the gene name column to match the cibersort output
  dplyr::select(., GeneSymbol = hgnc_symbol, everything()) # 13764 59
# remove samples not to be considered
cibersort <- cibersortx_reduced_mixture_adjusted %>%
  dplyr::select(., -SCP4060, -SCP4010, -SCP3492)

# What is the intersection of gene names between the two matrices
# common_strings <- intersect(cibersort$GeneSymbol, tpm$GeneSymbol) # no duplicated gene names
# number_common_strings <- length(common_strings)
# print(number_common_strings) # 12349

# rather than look for the intersection between the two expression tables I need to subset both by the genes in the signature matrix

# subset each table into only genes in the signature matrix by gene name
intersect_cibersort_adjusted_mixture <- subset(cibersort, GeneSymbol %in% cibersortx_reduced_sigMatrix$GeneSymbol) %>%
  melt() %>%
  dplyr::select(., GeneSymbol, samplename = variable, cibersort = value) # 753289 3

intersect_tpm <- subset(tpm, GeneSymbol %in% cibersortx_reduced_sigMatrix$GeneSymbol) %>%
  melt() %>%
  dplyr::select(., GeneSymbol, samplename = variable, tpm = value) #716242 3
# join the tables for ggplot - raw data
counts_to_plot <- dplyr::full_join(intersect_cibersort_adjusted_mixture, intersect_tpm, by = c("GeneSymbol", "samplename"))
ggplot(data = counts_to_plot,
       aes(x = tpm,
           y = cibersort,
           color = samplename)) +
  geom_point()

# lets log the counts so we can get a better view of the scatter
# save which columns are numeric
id_numeric <- sapply(counts_to_plot, is.numeric)
# make a copy of the raw data to transform
log2_counts_to_plot <- counts_to_plot
# transform the raw data
log2_counts_to_plot[, id_numeric] <- log2(counts_to_plot[, id_numeric] + 1)

# make and save the scatterplot - all samples in one plot
# png(filename=file.path(outdir, 'log2_adjusted_matrix_v_tpms.png'), width=800, height=800)
ggplot(data = log2_counts_to_plot,
       aes(x = tpm,
           y = cibersort,
           color = samplename)) +
  geom_point() 
# graphics.off()

# make and save one plot per sample
plots <- vector("list", length(unique(log2_counts_to_plot$samplename)))
for (i in seq_along(plots)) {
  samplename <- unique(log2_counts_to_plot$samplename)[i]
  subset <- log2_counts_to_plot[log2_counts_to_plot$samplename == samplename, ]
  
  # Create the plot object
  plots[[i]] <- ggplot(data = subset,
                       aes(x = tpm,
                           y = cibersort)) +
    geom_point() +
    ggtitle(samplename)
  
  # Save the plot as a PNG file
  png(filename = file.path(outdir, "cibersortx_mixture_matrix", paste0(samplename, "_log2_adjusted_matrix_v_tpms.png")),
      width = 800, height = 800)
  print(plots[[i]])
  dev.off()
}
```

## Create a new expression matrix based on the cell type proportion and the single cell signature matrix
### Start with one cell type
- Looks like there are only two genes in the signature matrix with values in only one cell type  

```{r}
# Identify rows with non-zero values in only one column of the signature matrix
rows_with_one_nonzero <- rowSums(cibersortx_reduced_sigMatrix[, -1] != 0) == 1
# Subset the original data frame based on the identified rows
subset_sig_matrix <- cibersortx_reduced_sigMatrix[rows_with_one_nonzero, ]
# Print the subsetted data frame
print(subset_sig_matrix)

# lets pull the two genes in question from the tpms
tpm_AADACL3 <- subset(tpm, GeneSymbol %in% "AADACL3")
tpm_GKN1 <- subset(tpm, GeneSymbol %in% "GKN1")

# and now lets calculate the expression for each gene based on the proportion of the ETV cell type
first_reconvolution_AADACL3 <- subset_sig_matrix %>% 
  subset(., GeneSymbol %in% "AADACL3") %>% melt()
  
second_reconvolution_AADACL3 <- subset_sig_matrix %>% 
  subset(., GeneSymbol %in% "AADACL3") %>%
  melt(., value.name = "signature_value", variable.names("GeneSymbol"))

reconvolution_GKN1

```

# Correlate gene expression with cell-type proportions
## Include genes that are both DE and in the cibersortx signature matrix  
- I will use the tpm matrix for the correlation as this is the table that is input into Cibersortx. Might not make any difference.  

```{r}

# put the hgnc symbol into the rownames for the tpm matrix
######tpm_bulkCountMatrix_all_genes %<>% tibble::column_to_rownames("hgnc_symbol")

# overlap between signature and DE genes - this is used to cut down the number of genes tested for correlation
sig_de_overlap <- subset(cibersortx_reduced_sigMatrix, GeneSymbol %in% topTable_male$hgnc_symbol)

# Initialize an empty data frame to store correlation results
final_correlation_results <- data.frame(Cell_Type = names(sig_de_overlap)[-1], stringsAsFactors = FALSE)

# loop through the genes and correlate with cell-type proportions
for (i in 1:nrow(sig_de_overlap)) {
  
  sig_de_overlap_ensbl <- dplyr::left_join(sig_de_overlap, topTable_male[, c("ensembl_gene_id", "hgnc_symbol")], by = c("GeneSymbol" = "hgnc_symbol")) %>%
    dplyr::select(., GeneSymbol, ensembl_gene_id, everything())
  
  GOI_ensbl <- paste0(sig_de_overlap_ensbl$ensembl_gene_id[i])
  GOI_hgnc <- paste0(sig_de_overlap_ensbl$GeneSymbol[i])
  
  cibersortx_proportion_rownames <- cibersortx_reduced_proportion %>%
    tibble::column_to_rownames("samplename")
  
  # subset the GOI tpm_bulkCountMatrix expression information for all samples
  tpm_GOI <- reshape2::melt(subset(tpm_bulkCountMatrix_all_genes, rownames(tpm_bulkCountMatrix_all_genes) %in% sig_de_overlap_ensbl$ensembl_gene_id[i])) %>% 
    as.data.frame() %>% 
    set_colnames(c("ensbl", "samplename", paste0(sig_de_overlap_ensbl$ensembl_gene_id[i]))) %>% 
    dplyr::select(., -ensbl) %>%
    dplyr::left_join(., cibersortx_reduced_proportion, by = "samplename") %>%
    dplyr::select(., -samplename)
  
  # Calculate correlations between gene expression and cell type proportions for each cell type
  correlations <- lapply(colnames(tpm_GOI)[-1], function(cell_type) {
    cor(tpm_GOI[[1]], tpm_GOI[[cell_type]])
  })
  
  # Combine correlations into a data frame
  correlation_results <- data.frame(names(tpm_GOI)[-1])
  
  # Add correlation results for the current gene as new columns
  colnames(correlation_results) <- paste(GOI_hgnc)
  correlation_results[] <- unlist(correlations)
  
  # Append the current gene's correlation results to the final data frame
  final_correlation_results <- cbind(final_correlation_results, correlation_results)
  
}


```
