---
title: "tpm_cpm_calc_20240418_readCounts"
author: "Melanie Smith"
date: "19 April 2024"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

This script converts the counts generated by `featureCounts` into transcripts per million (TPMs).  
Conversion to TPMs is required before we can compare the bulk counts with scRNAseq counts.  
In this case, the comparison will be used for cell type deconvolution.  

```{r}
library(IRanges)
library(GenomicFeatures)
library(edgeR)
library(readxl)
library(readr)
library(magrittr)

# set project directory
projectDir <- "/home/smit1924/preeclampsia_sex_chromosome_informed"
expt_name <- "20240418_readCounts"

genome_gtf_infile <- file.path(projectDir, "clearBoxCleanData/gencode.v29.annotation.gtf")
input_counts_20240418_female_readCounts_file <- file.path(projectDir, "clearBoxRawData/20240418_female_readCounts.txt")
input_counts_20240418_male_readCounts_file <- file.path(projectDir, "clearBoxRawData/20240418_male_readCounts.txt")
input_metadata_file <- file.path(projectDir, "clearBoxRawData/cleanSampleMetadata.csv")

outdir <- file.path(projectDir, "clearBoxCleanData")
output_20240418_readCounts_filelog2cpm_file <- file.path(outdir, "20240418_log2cpm.txt")

dedup_samples <- c("STP0148", "SCP3954", "STP1206", "STP0932", "SCP4733", "STP0596")
dir.create(outdir)

# set max digits
options(digits=3)

# set filtering criteria
filterCPM <- 2
numSamples <- 5
```

Before I can transform my raw counts to TPMs I need to calculate the gene lengths from the original .gtf file.   

```{r}

# Try it
# Specifics of calculating rpkm: https://support.bioconductor.org/p/64585/

# 1. Get gene lengths
# https://www.biostars.org/p/83901/
# Or using GTFtools, but keep chromosome limitations in mind:
# http://genomespot.blogspot.com/2019/01/using-gtf-tools-to-get-gene-lengths.html
# GTFtools offers flexibility about how to calculate the gene length (median, mean, etc.)

# Note, this is the gtf used in the RNA-Seq pipeline to annotate hg38
txdb <- GenomicFeatures::makeTxDbFromGFF(file = file.path(genome_gtf_infile), format = "gtf")
# Collect exons per gene id
exon_list_per_gene <- GenomicFeatures::exonsBy(txdb, by = "gene")
exon_list_per_gene

```

```{r}

# For each gene, reduce all exons to a set of non-overlapping exons, calculate their length (widths) and sum them
exonic_gene_sizes <- 
  sum(IRanges::width(IRanges::reduce(exon_list_per_gene))) %>% 
  tibble::as_tibble(rownames = NA) %>% 
  tibble::rownames_to_column(var = "gene") %>%
  dplyr::select(., gene, length = value)
exonic_gene_sizes

# calculate the number of unique genes in the "gene" column of the exonic_gene_sizes data frame.
dim(exonic_gene_sizes) #58721 2
dim(exonic_gene_sizes[!duplicated(exonic_gene_sizes$gene), ]) #58721 2
# at this point there are no duplicated ensembl ids

```

## Import metadata
- import the metadata then drop samples not yet deduplicated

```{r}
## Metadata
metadata <- read_csv(file = input_metadata_file)
```

## Import count files
### Male

```{r import gene counts - male}

# import the counts table
# counts are annotated to GRCh38
rawCounts_male <- read.delim(file = input_counts_20240418_male_readCounts_file) %>%
  as.data.frame()
# tidy up the column names
colnames(rawCounts_male) <- gsub("X.scratch.user.smit1924.20230608_male_grch38.aligned_data.|_GRCh38_Aligned.sortedByCoord.out.bam|_T.*|_",
                                 "",
                                 colnames(rawCounts_male))
# remove the genes from the PAR_Y (all have zero counts) they look like "ENSG00000185203.12_PAR_Y"
rawCounts_male %<>% dplyr::filter(., !grepl("PAR_Y",Geneid))
dim(rawCounts_male) #58686 24
# calculate the number of unique ensembl ids
dim(rawCounts_male[!duplicated(rawCounts_male$Geneid), ]) #58676 24
# at this point there are no dupkicated ensembl ids
```

### Female

```{r import gene counts - male}

# import the counts table
# counts are annotated to GRCh38
rawCounts_female <- read.delim(file = input_counts_20240418_female_readCounts_file) %>%
  as.data.frame()
# tidy up the column names
colnames(rawCounts_female) <- gsub("X.scratch.user.smit1924.20230608_female_grch38.aligned_data.|_GRCh38_Aligned.sortedByCoord.out.bam|_T.*|_",
                                 "",
                                 colnames(rawCounts_female))
# remove the genes from the PAR_Y (all have zero counts) they look like "ENSG00000185203.12_PAR_Y"
rawCounts_female %<>% dplyr::filter(., !grepl("PAR_Y",Geneid))
dim(rawCounts_female) #58686 39
# calculate the number of unique ensembl ids
dim(rawCounts_female[!duplicated(rawCounts_female$Geneid), ]) #58676 39
# at this point there are no dupkicated ensembl ids
```

## Combine counts data into a single df

```{r combine counts, echo=FALSE}

# test to make sure the rownames are the same
identical(rownames(rawCounts_male), rownames(rawCounts_female))

# combine the counts tables
rawCounts <- dplyr::left_join(rawCounts_male,
                              rawCounts_female,
                              by = "Geneid")
dim(rawCounts) # 58686 62

rownames(rawCounts) <- NULL # Resets numbers to start at one after omitting first four rows loading data
colnames(rawCounts)[1] <- "gene_id" # rename the gene column to match Nick's code

```

```{r}

# 2. Process raw count data in edgeR again, as per melanieRawCountsProcessing_v2.0.3.rmd
# Add gene length to expression data
raw_data <- rawCounts %>%
  dplyr::mutate(Length = exonic_gene_sizes$length[match(Geneid, exonic_gene_sizes$gene)])
# move the ensemble ids into the rownames and check table diminsions
raw_data <- raw_data %>% 
  tibble::column_to_rownames("Geneid")
head(raw_data)
dim(raw_data) # 58676 62

# establish the DGEList object
y <- edgeR::DGEList(counts = raw_data %>% dplyr::select(-Length),
                    genes = raw_data %>% dplyr::select(Length))

y
y$genes # A Length column is present 
dim(y) # 58676 61

head(y$counts) # Gene names are set as rownames
y$samples$lib.size # 25840016 14146199  9271133 23206041 13993327  4960561 11548145  9827585 11107194 13036244 19704650 10137751 18870585 25998630 28517294 14346006 20252292 10373457 10308117 10562185  9780885 11881306 10633676 27939519 15108240 21264765 10803912 10363357 12391073  6540180  8510134 11019145 22264896  9005622  7363172 17951449 12148454 22558944 19120886  9762863 21420996 21407095 19016558  7325552  9408056 12983552 24634481 23349226 14705521 25946757 18464833 23957877 27043669 10515540 13759070 20307551 19362575 10604993  9093579 12845802 21051760


# Identify genes expressed at > 1 cpm in at least one sample
keep <- rowSums(cpm(y) > filterCPM) >= numSamples
table(keep) # Keeping 13777 genes

# Apply filter
y <- y[keep, , keep.lib.sizes = FALSE]
dim(y) # 13777  61
y$samples$lib.size # 25731368 14109476  9235232 23113704 13947351  4943463 11521143  9768451 11076947 13008609 19636749 10113294 18825689 25941248 28421043 14305351 20173817 10339107 10284346 10530964  9755775 11856046 10607783 27873480 15031246 21207524 10776880 10319452 12355039  6499956  8476777 10981939 22206227  8962473  7334073 17868278 12119455 22494048 19076618  9731313 21349533 21351772 18966245  7296636  9389285 12948595 24578641 23298271 14630492 25892606 18422524 23885709 26963928 10481274 13720697 20251871 19318938 10580329  9060849 12813609 21010050

# TMM normalisation 
y$samples$norm.factors # currently all normalisation factors == 1
y <- calcNormFactors(y, method = "TMM")
y$samples$norm.factors # 1.112 0.921 1.019 1.005 0.855 1.058 1.030 1.078 1.109 0.910 1.130 0.993 1.039 0.901 0.956 1.025 1.031 1.154 1.019 0.932 0.901 0.973 0.924 1.014 1.169 0.992 1.047 1.010 1.004 0.962 1.050 0.959 0.954 1.084 1.200 1.001 1.015 1.052 0.969 0.880 0.948 1.027 1.090 0.971 0.897 0.994 0.929 0.946 1.180 0.942 0.974 1.015 0.954 0.921 0.935 1.059 0.946 0.986 1.065 0.977 0.973

# Now calculate RPKM: https://support.bioconductor.org/p/64585/
rpkm <- rpkm(y)

# And TPM: https://www.biostars.org/p/388584/#394579
tpm <- t( t(rpkm) / colSums(rpkm) ) * 1e6
# These TPM values now incorporate the TMM factors

```

## Generate and save a table of log2 cpms before adding the gene level information

```{r}
# generate the logged cpms
log2cpms_no_gene_names_20240418 <- edgeR::cpm(y, log = TRUE)

# save a copy of the logged cpms
log2cpms_no_gene_names_20240418 %>%
  as.data.frame() %>%
  # move the rownames into a column
  tibble::rownames_to_column("Geneid") %>%
  # save log2 cpm count file
  write.table(file = output_20240418_readCounts_filelog2cpm_file,
              quote = FALSE,
              sep = "\t",
              col.names = TRUE,
              row.names = FALSE)
```