---
title: "cibersortx_reduced_sigMatrix_compare"
author: "Melanie Smith"
date: "29 April 2024"
output: html_document
---

The aim of this script is to generate scatter plots to help us determine if CibersortX is accurately deconvoluting the deduplicated placenta data.  

```{r setup, include=FALSE}
knitr::opts_chunk$set(
    echo = TRUE,
    message = FALSE,
    warning = FALSE,
    cache = FALSE,
    fig.show='hold',
    fig.align = "center",
    results = "hide"
)

```

# Load required libraries

```{r load Libraries}
library(plyr)
library(reshape2)
library(dplyr)
library(tidyr)
library(stringr)
library(edgeR)
library(readxl)
library(readr)
library(magrittr)
library(tibble)
library(ggplot2)
library(ggbeeswarm)
library(ggrepel)
library(ggExtra)
library(pheatmap)
# Parallel optimisation
library(foreach)
library(doParallel)
# samtools
library(Rsamtools)
```

# Project directories and input/output files

```{r}

# set project directory
projectDir <- "/home/smit1924/preeclampsia_sex_chromosome_informed"
# set experiment name
expt_name <- 'deduplicated_bams'


# set the output directory
outdir <- file.path(projectDir, "deduplicated_clearBox/output/sigMatrixScatterPlots")
dir.create(outdir) # create the output directory if it doesn't already exist

# vector of deduplicated samplenames
dedup_samples <- c("STP0148", "SCP3954", "STP1206", "STP0932", "SCP4733", "STP0596")

# set max digits
options(digits=3)

# user generated functions
`%notin%` <- Negate(`%in%`)
```

# Input files for figure

```{r}
# tpm file used as input for the cibersortx deconvolution
input_tpm_12_sample_combined_bulkCountMatrix_file <- file.path(projectDir, "deduplicated_clearBox/output/tpm_12_sample_combined_bulkCountMatrix.txt")

# all files calculated using the 6 original (standard bams) and 6 deduplicated (deduplicated bams) tpms
## Cell type proportions (includes RMSE and correlation)
cibersortx_reduced_proportion_file <- file.path(projectDir, "deduplicated_clearBox/readCounts/CIBERSORTx_Adjusted.txt")

## Signature matrix after batch correction
cibersortx_reduced_sigMatrix_file <- file.path(projectDir, "deduplicated_clearBox/readCounts/CIBERSORTx_sigmatrix_Adjusted.txt")

## Predicted gene expression based on the proportions and new signature matrix
cibersortx_reduced_mixture_adjusted_file <- file.path(projectDir, "deduplicated_clearBox/readCounts/CIBERSORTx_Mixtures_Adjusted.txt")
```

## Additional input file

```{r}
# ## Genome files for v29 gencode (in case we want to add additional genome information)
# input_genome_gtf_file <- file.path(projectDir, "clearBoxCleanData/gencode.v29.annotation.gtf")
# input_genome_gff3_file <- file.path(projectDir, "clearBoxCleanData/gencode.v29.chr_patch_hapl_scaff.annotation.gff3")
# 
# ## Metadata file (in case we want to add information to a figure)
# input_metadata_file <- file.path(projectDir, "clearBoxRawData/cleanSampleMetadata.csv")
# 
# ## Deduplicated counts
# input_counts_deduplicated_file <- file.path(projectDir, "deduplicated_clearBox/readCounts/20240429_deduplicated_s2_readCounts.txt")
# input_dedup_no_split_ensembl_six_log2cpm_file <- file.path(projectDir, "deduplicated_clearBox/output", "dedup_no_split_ensembl_six_log2cpm.txt")
# 
# input_deduplicated_six_cpm_file <- file.path(projectDir, "deduplicated_clearBox/output", "deduplicated_six_cpm.txt")
# input_deduplicated_six_log2cpms_yes_gene_names_file <- file.path(projectDir, "deduplicated_clearBox/output", "deduplicated_six_log2cpms_yes_gene_names.txt")
# input_dedup_no_split_ensembl_six_log2cpm_file <- file.path(projectDir, "deduplicated_clearBox/output", "dedup_no_split_ensembl_six_log2cpm.txt")

```

# Import all input data tables
## Metadata

```{r}
## Metadata
# metadata <- read_csv(file = input_metadata_file)
# dedup_metadata <- dplyr::filter(metadata, samplename %in% dedup_samples)
```

## Deduplicated tpm bulk count matrix
- tpm file used as input for the cibersortx deconvolution  

```{r}
tpm_12_sample_combined_bulkCountMatrix <- read.table(file = input_tpm_12_sample_combined_bulkCountMatrix_file,
                                                 header = TRUE,
                                                 sep = "\t")
dim(tpm_12_sample_combined_bulkCountMatrix) # 13788 13
table(is.na(tpm_12_sample_combined_bulkCountMatrix))
table(duplicated(tpm_12_sample_combined_bulkCountMatrix$hgnc_symbol))

# there are no NAs and 14 duplicate hgnc_symbols
## lets take a look at the duplicate rows
duplicate_rows <- tpm_12_sample_combined_bulkCountMatrix[duplicated(tpm_12_sample_combined_bulkCountMatrix$hgnc_symbol) | duplicated(tpm_12_sample_combined_bulkCountMatrix$hgnc_symbol, fromLast = TRUE),]

```

## Cibersortx infiles

```{r}
# CIBERSORTx signature matrix and proportion files - reduced cell types
cibersortx_reduced_proportion <- read.table(file = cibersortx_reduced_proportion_file,
                                            header = TRUE,
                                            sep = "\t") %>%
  dplyr::select(., samplename = Mixture, everything()) %>%
  dplyr::select(., -'P.value', -'Correlation', -'RMSE')

# CIBERSORTx proportion file with RMSE  - reduced cell types
cibersortx_reduced_proportion_rmse <- read.table(file = cibersortx_reduced_proportion_file,
                                                 header = TRUE,
                                                 sep = "\t") %>%
  dplyr::select(., samplename = Mixture, everything())

cibersortx_reduced_sigMatrix <- read.table(file = cibersortx_reduced_sigMatrix_file,
                                           header = TRUE,
                                           sep = "\t")
# remove rows that contain all zero counts in the signature matrix
cibersortx_reduced_sigMatrix <- cibersortx_reduced_sigMatrix[rowSums(cibersortx_reduced_sigMatrix[,-1])>0,]
dim(cibersortx_reduced_sigMatrix)
# create a new matrix for the scaled data
cibersortx_reduced_sigMatrix_scaled <- cibersortx_reduced_sigMatrix  %>%
  `rownames<-`(NULL) %>%
  as.data.frame() %>%
  tibble::column_to_rownames("GeneSymbol") %>%
  t() %>%
  scale(., center = TRUE,
        scale = TRUE) %>%
  t()

# CIBERSORTx adjusted mixture - reduced cell types - this is the expression predicted by cibersort given the cell-type proportions and the signature matrix
cibersortx_reduced_mixture_adjusted <- read.table(file = cibersortx_reduced_mixture_adjusted_file,
                                                  header = TRUE,
                                                  sep = "\t") %>%
  # drop the "random1/2/3 etc" columns
  dplyr::select(., -starts_with("random"))
table(duplicated(cibersortx_reduced_mixture_adjusted$GeneSymbol))
# there are no duplicated GeneSymbol (same as hgnc_symbol but the Campbell paper calls the column GeneSymbol)

```

## Compare the ensembl IDs between each of the input tables

```{r}
## make a new table that contains the ensemble ID and table of origin for each input table

## Input tpm_12_sample_combined_bulkCountMatrix
## This file contains the 6 paired samples with hgnc
tpm_12 <- tpm_12_sample_combined_bulkCountMatrix %>%
  dplyr::select(., hgnc_symbol) %>%
  dplyr::mutate(tpm_12 = rep("tpm_12", nrow(.)))

## Input cibersort_reduced_sigMatrix
sigMat <- cibersortx_reduced_sigMatrix %>%
  dplyr::select(., hgnc_symbol = GeneSymbol) %>%
  dplyr::mutate(sigMat = rep("sigMat", nrow(.)))
  
## input cibersort_reduced_mixture_adjusted
mixMat <- cibersortx_reduced_mixture_adjusted %>%
  dplyr::select(., hgnc_symbol = GeneSymbol) %>%
  dplyr::mutate(mixMat = rep("mixMat", nrow(.)))

join_tables <- dplyr::full_join(tpm_12, sigMat, by = "hgnc_symbol") %>%
  dplyr::full_join(., mixMat, by = "hgnc_symbol")

```


# Predicted gene expression
## Scatter plot of the adjusted mixture matrix (cibersortx prediction) versus the sample tpms

```{r}

# there are 3 genes in the tpm matrix with duplicated gene names as an artifact of the original join - delete one and move on with life

# Find the indices of the first occurrence of each unique gene name
first_indices <- !duplicated(tpm_12_sample_combined_bulkCountMatrix$hgnc_symbol)

# Subset the matrix to keep only the rows with the first occurrence of duplicated gene names
tpm <- tpm_12_sample_combined_bulkCountMatrix[first_indices, ]
dim(tpm) # 19764 13

# subset each table to include only genes in the signature matrix by gene name
intersect_cibersort_adjusted_mixture <- subset(cibersortx_reduced_mixture_adjusted, GeneSymbol %in% cibersortx_reduced_sigMatrix$GeneSymbol) %>%
  melt() %>%
  dplyr::select(., GeneSymbol, samplename = variable, cibersort = value) # 223260 3

intersect_tpm <- subset(tpm, hgnc_symbol %in% cibersortx_reduced_sigMatrix$GeneSymbol) %>%
  melt() %>%
  dplyr::select(., GeneSymbol = hgnc_symbol, samplename = variable, tpm = value) # 21960 3

# join the tables for ggplot - raw data
counts_to_plot <- dplyr::full_join(intersect_cibersort_adjusted_mixture, intersect_tpm, by = c("GeneSymbol", "samplename"))
# lets log the counts so we can get a better view of the scatter
# save which columns are numeric
id_numeric <- sapply(counts_to_plot, is.numeric)
# make a copy of the raw data to transform
log2_counts_to_plot <- counts_to_plot
# Replace negative values with zero
log2_counts_to_plot[, id_numeric][log2_counts_to_plot[, id_numeric] < 0] <- 0
# transform the raw data
log2_counts_to_plot[, id_numeric] <- log2(log2_counts_to_plot[, id_numeric] + .01)

ggplot(data = log2_counts_to_plot,
       aes(x = tpm,
           y = cibersort,
           color = samplename)) +
  geom_point() +
  xlab("tpm (log2 + 0.01)") + 
  ylab("cibersortx (log2 + 0.01)") +
  theme_bw()

```

## Scatterplots per cell-type per sample
- after talking with Kat we decided we needed more resolution in the tpm v mixture matrix scatter plots.  
- to this end I will make new scatter plots, one per cell type per sample.  
- each plot will have the non-zero signature genes graded by expression value.  

```{r, eval=FALSE}

# lets create the new folders we need to hold all the plots
# each folder will be named by cell type from the proportion table
output_directory_names <- colnames(cibersortx_reduced_proportion[,-1])
name <- output_directory_names[1]
# Create directories based on column names
for (name in output_directory_names) {
  dir.create(path = paste0(outdir, "/", name), showWarnings = FALSE)  # Create directory with column name
}

## make and save one plot per cell type per sample
## Lets make an empty list to put plots in
plots <- vector("list", length(unique(log2_counts_to_plot$samplename)))
## Scale the signature matrix
scaled_sigMatrix <- cibersortx_reduced_sigMatrix_scaled  %>%
  as.data.frame() %>%
  tibble::rownames_to_column("GeneSymbol")

# Find the maximum value in the dataset for the axis length
max_value <- max(max(log2_counts_to_plot$tpm), max(log2_counts_to_plot$cibersort))
# Find the minimum value in the dataset for the axis length
min_value <- min(min(log2_counts_to_plot$tpm), min(log2_counts_to_plot$cibersort))

for (i in seq_along(plots)) {
  samplename <- unique(log2_counts_to_plot$samplename)[i]
  subset <- log2_counts_to_plot[log2_counts_to_plot$samplename == samplename, ]
  
  for (col_name in names(cibersortx_reduced_sigMatrix)[-1]) {
    # get the signature matrix values and combine with the tpm and cibersort values from subset
    subset_with_sigMatrix_value <- scaled_sigMatrix[scaled_sigMatrix[, col_name] >= 2, c("GeneSymbol", col_name)] %>%
      dplyr::left_join(., subset, by = "GeneSymbol")
    
    # Get the RMSE value corresponding to the current samplename
    rmse_value <- cibersortx_reduced_proportion_rmse[cibersortx_reduced_proportion_rmse$samplename == samplename, "RMSE"]
    # Get the Correlation value corresponding to the current samplename
    correlation_value <- cibersortx_reduced_proportion_rmse[cibersortx_reduced_proportion_rmse$samplename == samplename, "Correlation"]
    # Get the cell-type proportion value corresponding to the current samplename
    cell_type_proportion <- cibersortx_reduced_proportion_rmse[cibersortx_reduced_proportion_rmse$samplename == samplename, col_name]
    
    # Create the plot object
    p <- ggplot(data = subset,
                         aes(x = tpm,
                             y = cibersort)) +
      geom_point() +
      # Add line of best fit
      # geom_smooth(method = "lm", se = FALSE) +
      # set x and y axis limits
      xlim(min_value, max_value) +
      ylim(min_value, max_value) +
      
      # Add x=y line
      geom_abline(intercept = 0,
                  slope = 1,
                  color = "red",
                  linetype = "dashed") +
      
      # add the RMSE value in the top left corner
      geom_text(x = min(subset$tpm),
                y = max(subset$cibersort),
                label = paste("RMSE:", round(rmse_value, 2)),
                hjust = -1,
                color = "black") +
      # add the correlation
      geom_text(x = min(subset$tpm),
                y = max(subset$cibersort -0.5),
                label = paste("Correlation:", round(correlation_value, 2)),
                hjust = -1,
                color = "black") +
      # add the proportion of cell-type
      geom_text(x = min(subset$tpm),
                y = max(subset$cibersort -1),
                label = paste("Cell-Type Proportion:", round(cell_type_proportion, 2)),
                hjust = -1,
                color = "black") +
      
      # add a second layer of points coloured by sigMatrix value
      geom_point(data = subset_with_sigMatrix_value,
                 aes_string(x = "tpm",
                            y = "cibersort",
                            colour = col_name)) +
      
      # add a title to the plot
      ggtitle(paste(samplename, col_name, sep = " ")) +
      
      scale_color_gradient(name = paste("Cibersort SigMatrix (log2 ", col_name, ")", sep = ""),
                           low = "white",
                           high = "red") +
      theme_bw() +
      theme(legend.position = "bottom")
    
    
      # Add marginal plots
    plots[[i]] <- ggExtra::ggMarginal(p, type = "boxplot")
    
    
    # Save the plot as a PNG file
    png(filename = file.path(outdir, col_name, paste0(samplename, "_log2_adjusted_matrix_v_tpms.png")),
        width = 800, height = 800)
    print(plots[[i]])
    dev.off()
  }
}

```

```{r}
library(pheatmap)

# Select the columns to include in heatmap  
data <- cibersortx_reduced_proportion_rmse[,c(2:21)]

# Add sample names and annotation columns
rownames(data) <- cibersortx_reduced_proportion_rmse$samplename 
data <- cbind(data, cibersortx_reduced_proportion_rmse[,c("Correlation","RMSE")])

# Create the heatmap
pheatmap(t(data[c("STP0148", "STP0148_dedup", "SCP3954", "SCP3954_dedup", "STP1206", "STP1206_dedup", "STP0932", "STP0932_dedup", "SCP4733", "SCP4733_dedup", "STP0596", "STP0596_dedup"), 1:20]),
         annotation_col = data[,c("Correlation","RMSE")],
         cluster_cols = FALSE,
         cluster_rows = FALSE)

ggplot(cibersortx_reduced_proportion_rmse,
       aes(x = RMSE,
           y = Correlation)) +
  geom_point()

```

