---
title: "compare_STP0148"
author: "Melanie Smith"
date: "19 April 2024"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(
    echo = TRUE,
    message = FALSE,
    warning = FALSE,
    cache = FALSE,
    fig.show='hold',
    fig.align = "center",
    results = "hide"
)

```

# Load required libraries

```{r load Libraries}
library(plyr)
library(reshape2)
library(dplyr)
library(tidyr)
library(stringr)
library(edgeR)
library(readxl)
library(readr)
library(magrittr)
library(tibble)
# plots
library(ggplot2)
library(ggbeeswarm)
library(ggrepel)
library(ggExtra)
library(pheatmap)
library(corrplot)
library(heatmaply)
library(htmlwidgets)
# Parallel optimisation
library(foreach)
library(doParallel)
# samtools
library(Rsamtools)



# set project directory
projectDir <- "/home/smit1924/preeclampsia_sex_chromosome_informed"
expt_name <- "STP0148_compareCounts"

# input files (non-counts)
input_genome_gtf_file <- file.path(projectDir, "clearBoxCleanData/gencode.v29.annotation.gtf")
input_metadata_file <- file.path(projectDir, "clearBoxRawData/cleanSampleMetadata.csv")
input_metadata_quality_file <- file.path(projectDir, "clearBoxCleanData/metadata_quality.csv")

# input files (counts)
input_sixSample_dedup_featureCounts_file <- file.path(projectDir, "STP0148_compareCounts/deduplicated_readCounts.txt")
input_STP0148_dedup_featureCounts_s2_file <- file.path(projectDir, "STP0148_compareCounts/s2_STP0148_dedup_readCounts.txt")
input_STP0148_STAR_geneCount_file <- file.path(projectDir, "STP0148_compareCounts/STP_0148_GRCh38_ReadsPerGene.out.tab")
input_sixSample_dedup_featureCounts_s2_file <- file.path(projectDir, "STP0148_compareCounts/s2_deduplicated_readCounts.txt")
input_allMale_featureCounts_s2_file <- file.path(projectDir, "STP0148_compareCounts/featureCounts_v2.0.3_male_readCounts.txt")

# set the output directory
outdir <- file.path(projectDir, "STP0148_compareCounts/outDirectory")


dedup_samples <- c("STP0148", "SCP3954", "STP1206", "STP0932", "SCP4733", "STP0596")

# set max digits
options(digits=3)

```

# Import all input data tables
## sixSample_dedup_featureCounts

```{r}

# metadata file
metadata_quality <- read_csv(file = input_metadata_quality_file)

# input counts; tidy up the column names; concat detail to the column name
sixSample_dedup_featureCounts <- read.table(file = input_sixSample_dedup_featureCounts_file,
                                                  header = TRUE,
                                                  sep = "\t") %>%
  as.data.frame()
# tidy up the column names
colnames(sixSample_dedup_featureCounts) <- gsub("X.scratch.user.smit1924.preeclampsia_sex_chromosome_informed.deduplicated_clearBox.deduplicated_bams.|_marked_duplicates.bam",
                                 "",
                                 colnames(sixSample_dedup_featureCounts))
# concat detail to the column name
sixSample_dedup_featureCounts %<>% rename_at(vars(starts_with("SC") | starts_with("ST")), ~paste0(., "_sixSample_dedup_featureCounts"))

# remove the genes from the PAR_Y (all have zero counts) they look like "ENSG00000185203.12_PAR_Y"
sixSample_dedup_featureCounts %<>% dplyr::filter(., !grepl("PAR_Y",Geneid))
dim(sixSample_dedup_featureCounts) # 58676 7
# calculate the number of unique ensembl ids
dim(sixSample_dedup_featureCounts[!duplicated(sixSample_dedup_featureCounts$Geneid), ]) # 58676 7
# at this point there are no dupkicated ensembl ids
```

## STP0148_dedup_featureCounts_s2 

```{r}

# import the count table
STP0148_dedup_featureCounts_s2 <- read.table(file = input_STP0148_dedup_featureCounts_s2_file,
                                                  header = TRUE,
                                                  sep = "\t") %>%
  as.data.frame()
# tidy up the column names
colnames(STP0148_dedup_featureCounts_s2) <- gsub("X.scratch.user.smit1924.preeclampsia_sex_chromosome_informed.deduplicated_clearBox.deduplicated_bams.|_marked_duplicates.bam",
                                 "",
                                 colnames(STP0148_dedup_featureCounts_s2))
# concat detail to the column name
STP0148_dedup_featureCounts_s2 %<>% rename_at(vars(starts_with("SC") | starts_with("ST")), ~paste0(., "_STP0148_dedup_featureCounts_s2"))

# remove the genes from the PAR_Y (all have zero counts) they look like "ENSG00000185203.12_PAR_Y"
STP0148_dedup_featureCounts_s2 %<>% dplyr::filter(., !grepl("PAR_Y",Geneid))
dim(STP0148_dedup_featureCounts_s2) # 58676 2
# calculate the number of unique ensembl ids
dim(STP0148_dedup_featureCounts_s2[!duplicated(STP0148_dedup_featureCounts_s2$Geneid), ]) # 58676 2
# at this point there are no dupkicated ensembl ids

```

## STP0148_STAR_geneCount 

```{r}

# import the count table
STP0148_STAR_geneCount <- read.table(file = input_STP0148_STAR_geneCount_file,
                                                  header = TRUE,
                                                  sep = "\t") %>%
  dplyr::select(., Geneid = N_unmapped, unstranded = X1521284, forward = X1521284.1, STP0148_STAR_geneCount = X1521284.2) %>%
  as.data.frame() %>%
  dplyr::select(., Geneid, STP0148_STAR_geneCount)

# Filter out rows starting with "N"
STP0148_STAR_geneCount <- STP0148_STAR_geneCount[!grepl("^N", STP0148_STAR_geneCount$Geneid), , drop = FALSE]

# remove the genes from the PAR_Y (all have zero counts) they look like "ENSG00000185203.12_PAR_Y"
STP0148_STAR_geneCount %<>% dplyr::filter(., !grepl("PAR_Y",Geneid))
dim(STP0148_STAR_geneCount) # 58676 2
# calculate the number of unique ensembl ids
dim(STP0148_STAR_geneCount[!duplicated(STP0148_STAR_geneCount$Geneid), ]) # 58676 2
# at this point there are no dupkicated ensembl ids

```

## sixSample_dedup_featureCounts_s2 

```{r}

# import the count table
sixSample_dedup_featureCounts_s2 <- read.table(file = input_sixSample_dedup_featureCounts_s2_file,
                                                  header = TRUE,
                                                  sep = "\t") %>%
  as.data.frame()
# tidy up the column names
colnames(sixSample_dedup_featureCounts_s2) <- gsub("X.scratch.user.smit1924.preeclampsia_sex_chromosome_informed.deduplicated_clearBox.deduplicated_bams.|_marked_duplicates.bam",
                                 "",
                                 colnames(sixSample_dedup_featureCounts_s2))
# concat detail to the column name
sixSample_dedup_featureCounts_s2 %<>% rename_at(vars(starts_with("SC") | starts_with("ST")), ~paste0(., "_sixSample_dedup_featureCounts_s2"))

# remove the genes from the PAR_Y (all have zero counts) they look like "ENSG00000185203.12_PAR_Y"
sixSample_dedup_featureCounts_s2 %<>% dplyr::filter(., !grepl("PAR_Y",Geneid))
dim(sixSample_dedup_featureCounts_s2) # 58676 7
# calculate the number of unique ensembl ids
dim(sixSample_dedup_featureCounts_s2[!duplicated(sixSample_dedup_featureCounts_s2$Geneid), ]) # 58676 7
# at this point there are no dupkicated ensembl ids


```

## allMale_featureCounts_s2 

```{r}

# import the count table
allMale_featureCounts_s2 <- read.table(file = input_allMale_featureCounts_s2_file,
                                                  header = TRUE,
                                                  sep = "\t") %>%
  as.data.frame()
# tidy up the column names
colnames(allMale_featureCounts_s2) <- gsub("X.scratch.user.smit1924.20230608_male_grch38.aligned_data.|_T.*|_",
                                 "",
                                 colnames(allMale_featureCounts_s2))
# concat detail to the column name
allMale_featureCounts_s2 %<>% rename_at(vars(starts_with("SC") | starts_with("ST")), ~paste0(., "_allMale_featureCounts_s2"))

# remove the genes from the PAR_Y (all have zero counts) they look like "ENSG00000185203.12_PAR_Y"
allMale_featureCounts_s2 %<>% dplyr::filter(., !grepl("PAR_Y",Geneid))
dim(allMale_featureCounts_s2) # 58676 24
# calculate the number of unique ensembl ids
dim(allMale_featureCounts_s2[!duplicated(allMale_featureCounts_s2$Geneid), ]) # 58676 24
# at this point there are no dupkicated ensembl ids

```


# combine all raw counts

```{r}
# test to make sure the rownames are the same
identical(sixSample_dedup_featureCounts$Geneid,
          STP0148_dedup_featureCounts_s2$Geneid)
identical(sixSample_dedup_featureCounts$Geneid,
          STP0148_STAR_geneCount$Geneid)
identical(sixSample_dedup_featureCounts$Geneid,
          sixSample_dedup_featureCounts_s2$Geneid)
identical(sixSample_dedup_featureCounts$Geneid,
          allMale_featureCounts_s2$Geneid)
# all Geneid are the same

# List of tables to be joined
tables <- list(sixSample_dedup_featureCounts,
               STP0148_dedup_featureCounts_s2,
               STP0148_STAR_geneCount,
               sixSample_dedup_featureCounts_s2,
               allMale_featureCounts_s2)

# Join tables iteratively
combinedRawCounts <- Reduce(function(x, y) dplyr::left_join(x, y, by = "Geneid"), tables)
dim(combinedRawCounts) # 58676 38

# Remove rows with row sums equal to zero in every column
combinedRawCounts <- combinedRawCounts[rowSums(combinedRawCounts[, -1]) != 0, ]
dim(combinedRawCounts) # 42299 38

```

# count correlation
```{r}

rownames(combinedRawCounts) <- NULL # Resets numbers to start at one
# t_combinedRawCounts <- data.frame(combinedRawCounts) %>% tibble::column_to_rownames("Geneid") %>%
#   t()
combinedRawCounts <- data.frame(combinedRawCounts) %>% tibble::column_to_rownames("Geneid")

correlation <- cor(combinedRawCounts)

# Create an interactive heatmap
heatmap_plot <- heatmaply(correlation, 
                          xlab = "Variables", ylab = "Variables",
                          main = "Correlation Matrix Heatmap",
                          width = 1000, height = 800)

# Save the heatmap plot as an HTML file
# saveWidget(heatmap_plot,
#            file.path(outdir, "heatmap_plot.html"),
#            selfcontained = TRUE)


# corrplot::corrplot(correlation,
#                    type="upper",
#                    order="hclust",
#                    col=RColorBrewer::brewer.pal(n=8, name="RdYlBu"))

```


# Subset out STP0148 from each table

```{r}
STP0148 <- dplyr::left_join(log2cpm_20230608 %>% dplyr::select(., Geneid, STP0148_20230608),
                            log2cpm_20240418 %>% dplyr::select(., Geneid, STP0148_20240418),
                            by = "Geneid") %>%
  dplyr::left_join(., dedup_no_split_ensembl_log2cpm %>% dplyr::select(Geneid, STP0148_dedup),
                   by = "Geneid")


ENSG00000183668_STP0148 <- STP0148[str_detect(STP0148$Geneid, "^ENSG00000183668"), ]

```

# Scatter plots of logged cpms

```{r}
# set the min and max axis values for the plots
# Get column names of numeric columns
numeric_cols <- sapply(STP0148, is.numeric)
# Find the maximum value for each numeric column
max_values <- sapply(STP0148[, numeric_cols],
                     max,
                     na.rm = TRUE)
# Find the overall maximum value
overall_max <- round(max(max_values,
                         na.rm = TRUE), 0)
# Find the minimum value for each numeric column
min_values <- sapply(STP0148[, numeric_cols],
                     min,
                     na.rm = TRUE)
# Find the overall maximum value
overall_min <- round(min(min_values,
                         na.rm = TRUE), 0)

exp <- "STP0148_20240418_v_STP0148_dedup"
png(filename = file.path(outdir, paste0(exp, "_compare_logCPMs.png")),
    width = 800, height = 800)
STP0148 %>%
  ggplot() +
  geom_point(aes(x = STP0148_20240418,
                 y = STP0148_dedup)) +
        # set x and y axis limits
      xlim(overall_min, overall_max) +
      ylim(overall_min, overall_max) +
        geom_abline(intercept = 0,
                  slope = 1,
                  color = "red",
                  linetype = "dashed") +
        # update main and axis labels
      labs(x = "STP0148_20240418 log2 cpm",
           y = "STP0148_dedup log2 cpms",
           title = exp)  +
  theme_bw()
dev.off()


```



