---
title: "compare_STP0148"
author: "Melanie Smith"
date: "19 April 2024"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(
    echo = TRUE,
    message = FALSE,
    warning = FALSE,
    cache = FALSE,
    fig.show='hold',
    fig.align = "center",
    results = "hide"
)

```

# Load required libraries

```{r load Libraries}
library(plyr)
library(reshape2)
library(dplyr)
library(tidyr)
library(stringr)
library(edgeR)
library(readxl)
library(readr)
library(magrittr)
library(tibble)
library(ggplot2)
library(ggbeeswarm)
library(ggrepel)
library(ggExtra)
library(pheatmap)
# Parallel optimisation
library(foreach)
library(doParallel)
# samtools
library(Rsamtools)

# set project directory
projectDir <- "/home/smit1924/preeclampsia_sex_chromosome_informed"
expt_name <- "deduplicated_bams"


outdir <- file.path(projectDir, "deduplicated_clearBox/output")

genome_gtf_infile <- file.path(projectDir, "clearBoxCleanData/gencode.v29.annotation.gtf")
gencodev29_gff_inFile <- file.path(projectDir, "clearBoxCleanData/gencode.v29.chr_patch_hapl_scaff.annotation.gff3")
input_metadata_file <- file.path(projectDir, "clearBoxRawData/cleanSampleMetadata.csv")
input_counts_deduplicated <- file.path(projectDir, "deduplicated_clearBox/readCounts/deduplicated_readCounts.txt")
dedup_cpm_file <- file.path(outdir, "cpm_dedup.txt")
dedup_log2cpm_file <- file.path(outdir, "log2cpm_dedup.txt")
dedup_no_split_ensembl_log2cpm_file <- file.path(outdir, "dedup_no_split_ensembl_log2cpm.txt")
log2cpm_20240418_file <- file.path(projectDir, "clearBoxCleanData/20240418_log2cpm.txt")
log2cpm_20230608_file <- file.path(projectDir, "clearBoxCleanData/20230608_log2cpms_no_gene_names.txt")

dge_list_obj_file <- file.path(projectDir, "noCovariate_output/DGEList_filtered_normalised.rds")
metadata_quality_file <- file.path(projectDir, "clearBoxCleanData/metadata_quality.csv")

dedup_samples <- c("STP0148", "SCP3954", "STP1206", "STP0932", "SCP4733", "STP0596")
dir.create(outdir)

# set max digits
options(digits=3)

```

# Import all input data tables

```{r}

# metadata file
metadata_quality <- read_csv(file = metadata_quality_file)

# deduplicated lcpms
log2cpm_20230608 <- read.table(file = log2cpm_20230608_file, header = TRUE, sep = "\t") %>%
  rename_at(vars(starts_with("SC") | starts_with("ST")), ~paste0(., "_20230608"))
dim(log2cpm_20230608) # 13686 62
# check for duplicated ensembl gene IDs in the row names
dim(log2cpm_20230608[!duplicated(log2cpm_20230608$Geneid), ]) # 13672 9
head(log2cpm_20230608)

# 20240418 featureCounts
log2cpm_20240418 <- read.table(file = log2cpm_20240418_file, header = TRUE, sep = "\t") %>%
  rename_at(vars(starts_with("SC") | starts_with("ST")), ~paste0(., "_20240418"))
dim(log2cpm_20240418) # 13777 62
# check for duplicated ensembl gene IDs in the row names
dim(log2cpm_20240418[!duplicated(log2cpm_20240418$Geneid), ]) # 13777 9
head(log2cpm_20240418)

# deduplicated lcpms
dedup_no_split_ensembl_log2cpm <- read.table(file = dedup_no_split_ensembl_log2cpm_file, header = TRUE, sep = "\t") %>%
  rename_at(vars(starts_with("SC") | starts_with("ST")), ~paste0(., "_dedup"))
dim(dedup_no_split_ensembl_log2cpm) # 12367 9
# check for duplicated ensembl gene IDs in the row names
dim(dedup_no_split_ensembl_log2cpm[!duplicated(dedup_no_split_ensembl_log2cpm$gene), ]) # 12367 9
head(dedup_no_split_ensembl_log2cpm)


```

# Subset out STP0148 from each table

```{r}
STP0148 <- dplyr::left_join(log2cpm_20230608 %>% dplyr::select(., Geneid, STP0148_20230608),
                            log2cpm_20240418 %>% dplyr::select(., Geneid, STP0148_20240418),
                            by = "Geneid") %>%
  dplyr::left_join(., dedup_no_split_ensembl_log2cpm %>% dplyr::select(Geneid, STP0148_dedup),
                   by = "Geneid")
```

# Scatter plots of logged cpms

```{r}
# set the min and max axis values for the plots
# Get column names of numeric columns
numeric_cols <- sapply(STP0148, is.numeric)
# Find the maximum value for each numeric column
max_values <- sapply(STP0148[, numeric_cols],
                     max,
                     na.rm = TRUE)
# Find the overall maximum value
overall_max <- round(max(max_values,
                         na.rm = TRUE), 0)
# Find the minimum value for each numeric column
min_values <- sapply(STP0148[, numeric_cols],
                     min,
                     na.rm = TRUE)
# Find the overall maximum value
overall_min <- round(min(min_values,
                         na.rm = TRUE), 0)

exp <- "STP0148_20240418_v_STP0148_dedup"
png(filename = file.path(outdir, paste0(exp, "_compare_logCPMs.png")),
    width = 800, height = 800)
STP0148 %>%
  ggplot() +
  geom_point(aes(x = STP0148_20240418,
                 y = STP0148_dedup)) +
        # set x and y axis limits
      xlim(overall_min, overall_max) +
      ylim(overall_min, overall_max) +
        geom_abline(intercept = 0,
                  slope = 1,
                  color = "red",
                  linetype = "dashed") +
        # update main and axis labels
      labs(x = "STP0148_20240418 log2 cpm",
           y = "STP0148_dedup log2 cpms",
           title = exp)  +
  theme_bw()
dev.off()


```

