---
title: "clearBoxDE - post deduplication"
author: "Melanie Smith"
date: "8 July 2024"
output: html_document
---

# Load required libraries
```{r load Libraries}
library(plyr)
library(reshape2)
library(dplyr)
library(stringr)
library(edgeR)
library(readxl)
library(readr)
library(magrittr)
library(tibble)
library(ggplot2)
library(ggbeeswarm)
library(ggrepel)
library(pheatmap)
```

```{r}

# set project directory
projectDir <- "/home/smit1924/preeclampsia_sex_chromosome_informed"
# projectDir <- "/Users/pillmaka/Desktop/Melanie/drive-download"

# input files
counts_geneID_file <- file.path(projectDir, "deduplicated_clearBox/readCounts/deduplicated_s2_readCounts.txt")

gene_info_file <- file.path(projectDir, "clearBoxCleanData/gencode_v29_gene_id_symbol_chr_biotype.csv")
metadata_file <- file.path(projectDir, "clearBoxRawData/cleanSampleMetadata.csv")

input_cibersortx_post_dedup_proportions_file <- file.path(projectDir, "deduplicated_clearBox/cibersortx_output/CIBERSORTx_Job10_Adjusted.txt")
input_cibersortx_sigMatrix_file <- file.path(projectDir, "cibersortx_reduced_sigMatrix/outdir/CIBERSORTx_sigmatrix_Adjusted.txt")

# set the output directory
outdir <- file.path(projectDir, "deduplicated_clearBox/output/57_sample_noCovariate")

# output files
outDesignMatrix_file <- file.path(outdir, "designMatrix.csv")
output_spreadsheet_file <- file.path(outdir, "combined_allTable.csv")
output_pValue_file <- file.path(outdir, "pValue_DE.csv")
output_log2CPM_file <- file.path(outdir, "log2CPM.csv")
output_dgelist_57_file <- file.path(outdir, "filt_norm_dgelist_57.rds")

# set figure files
output_librarySize_file <- file.path(outdir, "deduplicated_librarySizePlot.png")
output_mds_file <- file.path(outdir, "deduplicated_57_mds.png")

dir.create(outdir)

# set max digits
options(digits=3)

# set filtering criteria
filterCPM <- 2
numSamples <- 4


`%notin%` <- Negate(`%in%`)

```

# Import all input data tables
## Raw Counts
```{r}
# import the counts table
# counts are annotated to GRCh38
rawCounts_deduplicated <- read.delim(file = counts_geneID_file) %>%
  as.data.frame()
# tidy up the column names
colnames(rawCounts_deduplicated) <- gsub("X.media.sf_D_DRIVE.VM_Projects.20240619_deduplicateBams.deduplicated_data.|_marked_duplicates.bam|_T.*|_",
                                 "",
                                 colnames(rawCounts_deduplicated))

```

## Gene information and sample metadata
```{r}

## Metadata
metadata <- read_csv(file = metadata_file) %>%
  # drop STP0148 from the metadata
  dplyr::filter(., samplename != "STP0148")

## Gene ID versus GeneSymbol
gene_info <- read_csv(file = gene_info_file)

## Gene ID versus GeneSymbol
gene_info <- read_csv(file = gene_info_file)
gene_info <- gene_info[!duplicated(gene_info$ensembl_gene_id), ]

```


# Plot the raw library sizes

```{r}
data.frame(colSums(rawCounts_deduplicated[, -1])) %>%
  magrittr::set_colnames("lib.size") %>%
  tibble::rownames_to_column("samplename") %>%
  mutate(color = case_when(
    lib.size < 5000000 ~ "red",
    lib.size >= 5000000 & lib.size < 10000000 ~ "blue",
    lib.size >= 10000000 ~ "darkgreen"
  )) %>%
  dplyr::filter(!samplename %in% c("SCP3492", "SCP4010", "SCP4060", "STP0148")) %>%
  ggplot(aes(x = reorder(samplename, -lib.size), y = lib.size, fill = color)) +
  geom_bar(stat = "identity") +
  geom_hline(yintercept = 5000000, linetype = "dashed", color = "red") +
  geom_hline(yintercept = 10000000, linetype = "dashed", color = "darkgreen") +
  labs(y = "Library size (total number of mapped and quantified reads)",
       x = "Samplename") +
  scale_fill_identity() +
  coord_flip() +
  ggtitle("Deduplicated Library Size")
ggsave(filename = output_librarySize_file,
       width = 7,
       height = 7,
       units = "in",
       dpi = 150)

```

# Establish DGElist object

```{r}
# drop the samples not required for this analysis
counts <- rawCounts_deduplicated %>%
  tibble::column_to_rownames("Geneid")
counts <- counts[, metadata$samplename]
dim(counts)
# drop rows with zero in all columns
counts <- counts[rowSums(counts[])>0,] %>%
  tibble::rownames_to_column("ensembl_gene_id")
dim(counts)
# add the hgnc gene symbol, chromosome and biotype to the counts matrix
counts_hgnc <- dplyr::left_join(counts, gene_info, by = "ensembl_gene_id") %>%
  dplyr::select(., ensembl_gene_id, ensembl, hgnc_symbol, seqnames, gene_type, everything())
dim(counts_hgnc) # 43423 63

# drop rows with NAs introduced in the join
counts_hgnc_clean <- counts_hgnc %>%
  filter(!is.na(ensembl))
dim(counts_hgnc_clean) # 42645 63

# establish the DGEList
dge_list_obj <- DGEList(counts = counts_hgnc_clean %>%
                          tibble::column_to_rownames("ensembl") %>%
                          dplyr::select(., -hgnc_symbol, -ensembl_gene_id, -seqnames, -gene_type),
                        samples = metadata,
                        group = metadata$group,
                        genes = counts_hgnc_clean %>% dplyr::select(ensembl, hgnc_symbol, seqnames, gene_type))
dim(dge_list_obj)

```

# Filter biological noise and plot

```{r}
# remove low information genes with zero counts across all samples
dge_list_obj <- dge_list_obj[rowSums(dge_list_obj$counts) > 0, ]
dim(dge_list_obj)

# new df of unfiltered cpm for the reduced DGEList
preFilterCPM <- cpm(dge_list_obj)

# remove low expressed genes (filterCPM in numSamples PE)
keep.exprs <- rowSums(preFilterCPM > filterCPM) >= numSamples

# Perform the filtering step and recalculate the TMM normalisation factors for each library.
dge_list_obj <- dge_list_obj[keep.exprs,,keep.lib.sizes = FALSE]

```

# TMM normalisation
```{r}
# calculate normalisation factors and apply to the DGEList object
dge_list_obj <- calcNormFactors(dge_list_obj,
                                method = "TMM")

# # set the new factor levels
sex_outcome_levels <- c("F_Control", "M_Control", "F_PE", "M_PE")
dge_list_obj$samples$sex_outcome <- factor(dge_list_obj$samples$sex_outcome, levels=sex_outcome_levels)

## save the filtered and normalised DGEList object
saveRDS(dge_list_obj,
        file = output_dgelist_57_file)
```

# Plot MDS

```{r}
# define colours for each sex_outcome group
sex_outcome_levels <- levels(as.factor(dge_list_obj$samples$sex_outcome))
colors <- rainbow(length(sex_outcome_levels))
# create a colour vector that matches the rows in the DGElist object
col_vector <- colors[as.numeric(as.factor(dge_list_obj$samples$sex_outcome))]

# save as png
# png(file = output_mds_file,
#     width = 12,
#     height = 9,
#     units = "in",
#     res = 150)
# Create the MDS plot with customizations
limma::plotMDS(dge_list_obj,
               col=col_vector,
               labels=dge_list_obj$samples$samplename,
               main="MDS Plot 57 samples - deduplicated")
legend("topright",
       legend=sex_outcome_levels,
       col=colors, pch=20,
       xpd=TRUE,
       inset=c(0.01, -0.1))
# dev.off()
```

# Calculate log2CPM and save table as .csv

```{r}
log2CPM <- edgeR::cpm(dge_list_obj, log = TRUE)

# edgeR::cpm(dge_list_obj, log = TRUE) %>%
#   as.data.frame() %>%
#   tibble::rownames_to_column("ensembl") %>%
#   write_csv(., file = output_log2CPM_file,
#            col_names = TRUE)

```

# Differential Expression
- covariates are cell-type proportions

```{r}

#Create and write out design matrix
model_noCovariate <- '~0 + sex_outcome'
# establish the design matrix
design_sex_PE <- model.matrix(formula(model_noCovariate),
                              data = dge_list_obj$samples)
# make the column names a little nicer
colnames(design_sex_PE) <- c(sex_outcome_levels)

# save a csv of the design matrix
# write_csv(data.frame(design_sex_PE) %>%
#             tibble::rownames_to_column("samplename"),
#           file = outDesignMatrix_file,
#           col_names = TRUE)
```

```{r}
# perform the voom
voom_sex_PE <- voom(dge_list_obj, design_sex_PE, plot = TRUE)

# Check for non-estimable coefficients in the design matrix
non_est <- nonEstimable(design_sex_PE)
if (!is.null(non_est)) {
  warning("Non-estimable coefficients: ", paste(non_est, collapse = ", "))
}

# fit the linear model
fit_sex_PE <- lmFit(voom_sex_PE, design_sex_PE)
```

```{r}
# set the contrasts - this makes it easier to "see" what we're testing
contrast_sex_PE <- makeContrasts(female_PE = F_PE-F_Control,
                                 male_PE = M_PE-M_Control,
                                 levels = design_sex_PE)
contrast_sex_PE

# fit a linear regression to the contrast questions
contrast_fit_sex_PE <- contrasts.fit(fit_sex_PE, contrast_sex_PE)

# perform bayesian adjustment
contrast_fit_sex_PE <- eBayes(contrast_fit_sex_PE)

# grab the residuals
res_post_dedup <- residuals(fit_sex_PE,voom_sex_PE)
saveRDS(res_post_dedup,
        file = file.path(outdir, "residuals.rds"))

# grab the fitted values
fitted_post_dedup <- fitted(fit_sex_PE)
saveRDS(fitted_post_dedup,
        file = file.path(outdir, "fitted_values.rds"))

# check the coefficients
coefficients <- contrast_fit_sex_PE$coefficients
saveRDS(coefficients,
        file = file.path(outdir, "coefficients.rds"))
dim(contrast_fit_sex_PE$coefficients)
```

```{r}
# summary table of the sex PE fit
summary(decideTests(contrast_fit_sex_PE,
                    adjust.method = "fdr",
                    p.value = 0.05))

# all DE results for the female_PE comparison
allTable_female_PE <- topTable(contrast_fit_sex_PE,
                             coef = "female_PE",
                             n = Inf,
                             sort = "p")
# create output files and save
write_csv(allTable_female_PE,
          file=file.path(outdir, "allTable_female.csv"))

# write_csv(allTable_female_PE,
#           file="/media/sf_D_DRIVE/VM_Projects/20240723_compare_pre_post_dedup/allTable_female_post_dedup_cell_type_proportion.csv")


# all DE results for the male comparison
allTable_male_PE <- topTable(contrast_fit_sex_PE,
                             coef = "male_PE",
                             n = Inf,
                             sort = "p")
write_csv(allTable_male_PE,
          file=file.path(outdir, "allTable_male.csv"))

# write_csv(allTable_male_PE,
#           file="/media/sf_D_DRIVE/VM_Projects/20240723_compare_pre_post_dedup/allTable_male_post_dedup_cell_type_proportion.csv")

topTable_male_pe <- topTable(contrast_fit_sex_PE,
                             coef = "male_PE",
                             n = Inf,
                             sort = "p",
                             p = 0.05,
                             adjust.method = "fdr")

# write out a few summary tables
df <- left_join(allTable_female_PE %>% dplyr::select(-t, -P.Value, -B),
               allTable_male_PE %>% dplyr::select(-t, -P.Value, -B),
               by=c('hgnc_symbol', 'ensembl'),
          suffix=c("_F", "_M"))

coeff_and_sigma_df <- cbind(fit_sex_PE$coefficients, sigma=fit_sex_PE$sigma) %>%
  as.data.frame() %>%
  rownames_to_column(var="ensembl")

df <- left_join(coeff_and_sigma_df,
               df,
               by="ensembl")

write_csv(df %>% relocate(hgnc_symbol, .after=ensembl),
          file=output_spreadsheet_file)

df2 <- left_join(allTable_female_PE %>% dplyr::select(ensembl, hgnc_symbol, t, P.Value, adj.P.Val),
               allTable_male_PE %>% dplyr::select(ensembl, hgnc_symbol, t, P.Value, adj.P.Val),
               by=c('hgnc_symbol', 'ensembl'),
          suffix=c("_F", "_M"))

write_csv(df2 %>% relocate(hgnc_symbol, .after=ensembl),
          file=output_pValue_file)

```

## Sanity plot of DE genes

```{r}
# let's take a look at the DE genes with the highest logFC that also have a "decent" expression level

# row center the counts
# do this by subtracting the row(gene) mean from each count

# center with 'scale()'
# setting `scale = FALSE` subtracts the mean rather than zero centering
center_scale <- function(x) {
    scale(x, scale = FALSE)
}

# apply it
row_centered_counts <- center_scale(t(dge_list_obj$counts)) %>% 
  t()

# png(filename=file.path(outdir, "expression_DE_gene_CP_sexOutcome.png'), width=800, height=800)
# Gene of Interest (GOI)
GOI_PE_UP <- data.frame(ensbl=c("ENSG00000026508", "ENSG00000134184",
                                "ENSG00000149084", "ENSG00000163453",
                                "ENSG00000143995", "ENSG00000091039",
                                "ENSG00000134533"),
                        gene=c("CD44", "GSTM1", "HSD17B12", "IGFBP7", "MEIS1", "OSBPL8", "RERG"),
                        direction = "up")

GOI_PE_male_UP <- allTable_male_PE %>%
  arrange(desc(logFC)) %>%
  dplyr::select(., ensbl = ensembl, gene = hgnc_symbol)

GOI_PE_DOWN <- data.frame(ensbl=c("ENSG00000173917", "ENSG00000275993"),
                          gene=c("HOXB2", "SIK1B"),
                          direction = "down")

GOI_UP_DOWN <- rbind(GOI_PE_UP, GOI_PE_DOWN)

samples <- dge_list_obj$samples

counts <- edgeR::cpm(dge_list_obj, log = TRUE)

pd <- position_dodge(width = 0.5)

lcpm <- edgeR::cpm(dge_list_obj, log = TRUE)

## Print all of the plots at once
for (i in 1:nrow(GOI_PE_DOWN)) {
melt_lcpm <- melt(subset(row_centered_counts, rownames(row_centered_counts) %in% GOI_PE_DOWN$ensbl[i])) %>% 
  as.data.frame() %>% 
  set_colnames(c("ensbl", "samplename", "log2cpm")) %>% 
  # tibble::rownames_to_column("samplename") %>% 
  left_join(., dge_list_obj$samples[, c("samplename", "Outcome", "Sex")], by = "samplename") 
  
beeswarm <- ggplot(melt_lcpm, aes(x = Sex,
                                y = log2cpm,
                                fill = Outcome)) +
  stat_boxplot(geom="errorbar",
               position = pd,
               width = 0.2) +
  geom_boxplot(width = 0.5,
               position = pd) +
  stat_summary(fun = mean,
               geom = "point",
               size=2,
               shape = 5,
               position = position_dodge(0.5)) +
  geom_beeswarm(dodge.width = 0.6, cex = 2) +
  theme_bw(base_size = 18) +
  ylab(paste("Centered and Scaled Counts", GOI_PE_DOWN$ensbl[i], GOI_PE_DOWN$gene[i], sep = " - ")) +
  theme(axis.text.x = element_text(colour = "black")) +
  theme(legend.position = "bottom") +
      geom_text_repel(data = subset(melt_lcpm, Outcome == "PE" & Sex == "M"),
                    aes(label = samplename),
                    nudge_y = 0.2,
                    size = 3)
print(beeswarm)

  # Save the plot
  filename <- paste("/media/sf_D_DRIVE/VM_Projects/20240619_deduplicateBams/plots/Scatter_and_Box_Plot", GOI_PE_DOWN$ensbl[i], GOI_PE_DOWN$gene[i], ".png", sep = "_")
  ggsave(filename, plot = beeswarm, units = "in", width = 12, height = 12, dpi = 300)
  
}

# graphics.off()
```

## Volcano plot for visualisation of differential expression
### male

```{r volcano plots male}

decide <- decideTests(contrast_fit_sex_PE)
d2 <- topTable(contrast_fit_sex_PE, coef = 2, n = Inf, sort = "p") #[,c(2,6)]
d2$threshold <- 0
d2$threshold[0:nrow(filter(allTable_male_PE, adj.P.Val < 0.05))] <- 1
d2$threshold <- as.factor(d2$threshold)
d3 <- decideTests(contrast_fit_sex_PE) %>%
  as.data.frame()
d3 <- d3[2]
d4 <- left_join(tibble::rownames_to_column(d2), (tibble::rownames_to_column(d3)),
                       by = "rowname")
d4$male_PE <- as.factor(d4$male_PE)
d4$neg.log10FDR <- -log10(d4$adj.P.Val)
# set colours for the volcano plot
colour <- c("red", "black", "darkgreen")
FDR_FC_Decide <- decideTests(contrast_fit_sex_PE, lfc = 0)[,"male_PE"] %>%
  as.data.frame() %>%
  set_colnames("FDR_FC_Decide") %>%
  tibble::rownames_to_column() %>%
  left_join(., d4, by = "rowname")
FDR_FC_Decide$FDR_FC_Decide <- as.factor(FDR_FC_Decide$FDR_FC_Decide)

volcano_male <- ggplot(data = FDR_FC_Decide,
                  aes(x = logFC,
                      y = neg.log10FDR,
                      label = rowname,
                      colour = FDR_FC_Decide)) +
  geom_point(alpha=0.9, size=2.00) +
  xlab(expression(Log[2]*" Fold Change")) + ylab(expression("-ve "*Log[10]*" FDR")) +
  scale_color_manual(values = colour, name="mRNA\nRegulation",
                         breaks = c("-1", "0", "1"),
                         labels = c("Down-regulated", "Not Significant", "Up-regulated")) +
  theme(plot.title = element_text(hjust = 0.5)) +
  geom_hline(yintercept = -log10(.05), linetype = "dotted") +
  geom_vline(xintercept = c(-1, 1), linetype = "dotted") +
  theme_bw() +
  theme(text = element_text(size=18)) +
  theme(axis.text.x = element_text(colour = "black",
                                   face = "bold"),
        axis.text.y = element_text(colour = "black",
                                   face = "bold")) +
  theme(axis.title.x = element_text(colour = "black",
                                   face = "bold"),
        axis.title.y = element_text(colour = "black",
                                   face = "bold")) +
  geom_label_repel(data = dplyr::filter(FDR_FC_Decide, FDR_FC_Decide==1),
                  aes(label = hgnc_symbol),
                  fill = "white",
                  xlim = c(7, 0),
                  max.overlaps = Inf,
                  show.legend = FALSE) +
  geom_label_repel(data = dplyr::filter(FDR_FC_Decide, FDR_FC_Decide==-1),
                  aes(label = hgnc_symbol),
                  fill = "white",
                  xlim = c(-7, 0),
                  max.overlaps = Inf,
                  show.legend = FALSE) +
  theme(legend.position="bottom")

ggsave(plot = volcano_male,
       filename = file.path(outdir, "volcano_male.png"),
       units = "in",
       width = 12,
       height = 12,
       dpi = 300)


```

### female

```{r volcano plots female}

decide <- decideTests(contrast_fit_sex_PE)
d2 <- topTable(contrast_fit_sex_PE, coef = 1, n = Inf, sort = "p") #[,c(2,6)]
d2$threshold <- 0
d2$threshold[0:nrow(filter(allTable_female_PE, adj.P.Val < 0.05))] <- 1
d2$threshold <- as.factor(d2$threshold)
d3 <- decideTests(contrast_fit_sex_PE) %>%
  as.data.frame()
d3 <- d3[1]
d4 <- left_join(tibble::rownames_to_column(d2), (tibble::rownames_to_column(d3)),
                       by = "rowname")
d4$female_PE <- as.factor(d4$female_PE)
d4$neg.log10FDR <- -log10(d4$adj.P.Val)
# set colours for the volcano plot
colour <- c("red", "black", "darkgreen")
FDR_FC_Decide <- decideTests(contrast_fit_sex_PE, lfc = 0)[,"female_PE"] %>%
  as.data.frame() %>%
  set_colnames("FDR_FC_Decide") %>%
  tibble::rownames_to_column() %>%
  left_join(., d4, by = "rowname")
FDR_FC_Decide$FDR_FC_Decide <- as.factor(FDR_FC_Decide$FDR_FC_Decide)

volcano_female <- ggplot(data = FDR_FC_Decide,
                  aes(x = logFC,
                      y = neg.log10FDR,
                      label = rowname,
                      colour = FDR_FC_Decide)) +
  geom_point(alpha=0.9, size=2.00) +
  xlab(expression(Log[2]*" Fold Change")) + ylab(expression("-ve "*Log[10]*" FDR")) +
  scale_color_manual(values = colour, name="mRNA\nRegulation",
                         breaks = c("-1", "0", "1"),
                         labels = c("Down-regulated", "Not Significant", "Up-regulated")) +
  theme(plot.title = element_text(hjust = 0.5)) +
  geom_hline(yintercept = -log10(.05), linetype = "dotted") +
  geom_vline(xintercept = c(-1, 1), linetype = "dotted") +
  theme_bw() +
  theme(text = element_text(size=18)) +
  theme(axis.text.x = element_text(colour = "black",
                                   face = "bold"),
        axis.text.y = element_text(colour = "black",
                                   face = "bold")) +
  theme(axis.title.x = element_text(colour = "black",
                                   face = "bold"),
        axis.title.y = element_text(colour = "black",
                                   face = "bold")) +
  geom_label_repel(data = dplyr::filter(FDR_FC_Decide, FDR_FC_Decide==1),
                  aes(label = hgnc_symbol),
                  fill = "white",
                  xlim = c(7, 0),
                  max.overlaps = Inf,
                  show.legend = FALSE) +
  geom_label_repel(data = dplyr::filter(FDR_FC_Decide, FDR_FC_Decide==-1),
                  aes(label = hgnc_symbol),
                  fill = "white",
                  xlim = c(-7, 0),
                  max.overlaps = Inf,
                  show.legend = FALSE) +
  theme(legend.position="bottom")

ggsave(plot = volcano_female,
       filename = file.path(outdir, "volcano_female.png"),
       units = "in",
       width = 12,
       height = 12,
       dpi = 300)


```

# Heatmaps
```{r}
# Modify ordering of the clusters using clustering callback option
callback = function(hc, mat){
    sv = svd(t(mat))$v[,1]
    dend = reorder(as.dendrogram(hc), wts = sv)
    as.hclust(dend)
}

annotation_cols <- metadata[, c("samplename", "Sex", "Outcome")] %>%
  tibble::column_to_rownames("samplename")

scaled_exp <- scale(t(subset(log2CPM, rownames(log2CPM) %in% dplyr::filter(FDR_FC_Decide, threshold == 1)$ensembl)))

png(file=file.path(outdir, "heatmap_scaled_sig_AdjPVal_M.png"),
    width=12,
    height=8,
    units = "in",
    res = 150)
pheatmap(t(scaled_exp),
         annotation_col = annotation_cols,
         show_rownames = FALSE,
         show_colnames = TRUE,
         cutree_cols = 3,
         clustering_callback = callback)
dev.off()

```

# Session information
```{r session info}

sessionInfo()

```