---
title: "clearBoxDE - post deduplication"
author: "Melanie Smith"
date: "8 July 2024"
output: html_document
---

# Load required libraries
```{r load Libraries}
library(plyr)
library(reshape2)
library(dplyr)
library(stringr)
library(edgeR)
library(readxl)
library(readr)
library(magrittr)
library(tibble)
library(ggplot2)
library(ggbeeswarm)
library(ggrepel)
library(pheatmap)

# set project directory
projectDir <- "/home/smit1924/preeclampsia_sex_chromosome_informed"
# projectDir <- "/Users/pillmaka/Desktop/Melanie/drive-download"

# input files
counts_geneID_file <- file.path(projectDir, "deduplicated_clearBox/readCounts/deduplicated_s2_readCounts.txt")

gene_info_file <- file.path(projectDir, "clearBoxCleanData/gencode_v29_gene_id_symbol_chr_biotype.csv")
metadata_file <- file.path(projectDir, "clearBoxRawData/cleanSampleMetadata.csv")

# set the output directory
outdir <- file.path(projectDir, "deduplicated_clearBox/output/58_sample_noCovariate")

# output files
outDesignMatrix_file <- file.path(outdir, "designMatrix.csv")
output_spreadsheet_file <- file.path(outdir, "combined_allTable.csv")
output_pValue_file <- file.path(outdir, "pValue_DE.csv")
output_log2CPM_file <- file.path(outdir, "log2CPM.csv")
output_dgelist_58_file <- file.path(outdir, "filt_norm_dgelist_58.rds")
output_gene_wise_bcv_results_male_file <- file.path(outdir, "dedup_gene_wise_bcv_results_male.csv")
output_gene_wise_bcv_results_female_file <- file.path(outdir, "dedup_gene_wise_bcv_results_female.csv")

# set figure files
output_librarySize_file <- file.path(outdir, "deduplicated_librarySizePlot.png")
output_mds_file <- file.path(outdir, "deduplicated_58_mds.png")

dir.create(outdir)

# set max digits
options(digits=3)

# set filtering criteria
filterCPM <- 2
numSamples <- 5

```

# Import all input data tables
## Raw Counts
```{r}
# import the counts table
# counts are annotated to GRCh38
rawCounts_deduplicated <- read.delim(file = counts_geneID_file) %>%
  as.data.frame()
# tidy up the column names
colnames(rawCounts_deduplicated) <- gsub("X.media.sf_D_DRIVE.VM_Projects.20240619_deduplicateBams.deduplicated_data.|_marked_duplicates.bam|_T.*|_",
                                 "",
                                 colnames(rawCounts_deduplicated))

```

## Gene information and sample metadata
```{r}

## Metadata
metadata <- read_csv(file = metadata_file)

## Gene ID versus GeneSymbol
gene_info <- read_csv(file = gene_info_file)

## Gene ID versus GeneSymbol
gene_info <- read_csv(file = gene_info_file)
gene_info <- gene_info[!duplicated(gene_info$ensembl_gene_id), ]

```

# Plot the raw library sizes

```{r}
data.frame(colSums(rawCounts_deduplicated[, -1])) %>%
  rename(lib.size = colSums.rawCounts_deduplicated....1..) %>%
  tibble::rownames_to_column("samplename") %>%
  mutate(color = case_when(
    lib.size < 5000000 ~ "red",
    lib.size >= 5000000 & lib.size < 10000000 ~ "blue",
    lib.size >= 10000000 ~ "darkgreen"
  )) %>%
  dplyr::filter(!samplename %in% c("SCP3492", "SCP4010", "SCP4060")) %>%
  ggplot(aes(x = reorder(samplename, -lib.size), y = lib.size, fill = color)) +
  geom_bar(stat = "identity") +
  geom_hline(yintercept = 5000000, linetype = "dashed", color = "red") +
  geom_hline(yintercept = 10000000, linetype = "dashed", color = "darkgreen") +
  labs(y = "Library size (total number of mapped and quantified reads)",
       x = "Samplename") +
  scale_fill_identity() +
  coord_flip() +
  ggtitle("Deduplicated Library Size")
# ggsave(filename = output_librarySize_file,
#        width = 7,
#        height = 7,
#        units = "in",
#        dpi = 150)

```

# Establish DGElist object

```{r}
# drop the samples not required for this analysis
counts <- rawCounts_deduplicated %>%
  tibble::column_to_rownames("Geneid")
counts <- counts[, metadata$samplename]

# drop rows with zero in all columns
counts <- counts[rowSums(counts[])>0,] %>%
  tibble::rownames_to_column("ensembl_gene_id")

# add the hgnc gene symbol, chromosome and biotype to the counts matrix
counts_hgnc <- dplyr::left_join(counts, gene_info, by = "ensembl_gene_id") %>%
  dplyr::select(., ensembl_gene_id, ensembl, hgnc_symbol, seqnames, gene_type, everything())
dim(counts_hgnc) # 43423 63

# drop rows with NAs introduced in the join
counts_hgnc_clean <- counts_hgnc %>%
  filter(!is.na(ensembl))
dim(counts_hgnc_clean) # 42645 63

# establish the DGEList
dge_list_obj <- DGEList(counts = counts_hgnc_clean %>%
                          tibble::column_to_rownames("ensembl") %>%
                          dplyr::select(., -hgnc_symbol, -ensembl_gene_id, -seqnames, -gene_type),
                        samples = metadata,
                        group = metadata$group,
                        genes = counts_hgnc_clean %>% dplyr::select(ensembl, hgnc_symbol, seqnames, gene_type))

```

# Filter biological noise and plot

```{r}

# new df of unfiltered cpm for the reduced DGEList
preFilterCPM <- cpm(dge_list_obj)

# remove low expressed genes (filterCPM in numSamples PE)
keep.exprs <- rowSums(preFilterCPM > filterCPM) >= numSamples

# Perform the filtering step and recalculate the TMM normalisation factors for each library.
dge_list_obj <- dge_list_obj[keep.exprs,,keep.lib.sizes = FALSE]

dim(dge_list_obj)

```

# TMM normalisation
```{r}
# calculate normalisation factors and apply to the DGEList object
dge_list_obj <- calcNormFactors(dge_list_obj,
                                method = "TMM")

# # set the new factor levels
sex_outcome_levels <- c("F_Control", "M_Control", "F_PE", "M_PE")
dge_list_obj$samples$sex_outcome <- factor(dge_list_obj$samples$sex_outcome, levels=sex_outcome_levels)

## save the filtered and normalised DGEList object
# saveRDS(dge_list_obj,
#         file = output_dgelist_58_file)
```

# Plot MDS

```{r}
# define colours for each sex_outcome group
sex_outcome_levels <- levels(as.factor(dge_list_obj$samples$sex_outcome))
colors <- rainbow(length(sex_outcome_levels))
# create a colour vector that matches the rows in the DGElist object
col_vector <- colors[as.numeric(as.factor(dge_list_obj$samples$sex_outcome))]

# save as png
# png(file = output_mds_file,
#     width = 12,
#     height = 9,
#     units = "in",
#     res = 150)
# Create the MDS plot with customizations
limma::plotMDS(dge_list_obj, col=col_vector, labels=dge_list_obj$samples$samplename, main="MDS Plot 58 samples - deduplicated")
legend("topright", legend=sex_outcome_levels, col=colors, pch=20, xpd=TRUE, inset=c(0.01, -0.1))
# dev.off()
```
# Sex-specific BCV

```{r}
# vector of only male sample names
males <- dplyr::filter(dge_list_obj$samples, Sex == "M")[, "samplename"]
# subset DGEList to contain only males
y_male <- dge_list_obj[, males]
# calculate and plot BCV for male samples only
y_male <- estimateCommonDisp(y_male,
                             verbose = TRUE)
y_male <- estimateTrendedDisp(y_male)
y_male <- estimateTagwiseDisp(y_male)

# Extract gene-wise BCV values
gene_wise_bcv_male <- sqrt(y_male$tagwise.dispersion)

# Create a data frame with gene names and BCV values
bcv_table_male <- data.frame(
  Gene = rownames(y_male),
  BCV = gene_wise_bcv_male,
  AveExpr = y_male$AveLogCPM
)

# save the table to a CSV file:
# write.csv(bcv_table_male, file = output_gene_wise_bcv_results_male_file,
# row.names = FALSE)

# plot the BCVs
plotBCV(y_male,
        xlab = "Average log CPM",
        ylab = "Biological coefficient of variation - male",
        pch = 16,
        cex = 0.2,
        col.common = "red",
        col.trend = "blue",
        col.tagwise = "black")

# vector of only female sample names
females <- dplyr::filter(dge_list_obj$samples, Sex == "F")[, "samplename"]
# subset DGEList to contain only females
y_female <- dge_list_obj[, females]
# calculate and plot BCV for female samples only
y_female <- estimateCommonDisp(y_female,
                               verbose = TRUE)
y_female <- estimateTrendedDisp(y_female)
y_female <- estimateTagwiseDisp(y_female)
# Extract gene-wise BCV values
gene_wise_bcv_female <- sqrt(y_female$tagwise.dispersion)
# Create a data frame with gene names and BCV values
bcv_table_female <- data.frame(
  Gene = rownames(y_female),
  BCV = gene_wise_bcv_female,
  AveExpr = y_female$AveLogCPM
)

# save the table to a CSV file:
# write.csv(bcv_table_female, file = output_gene_wise_bcv_results_female_file,
# row.names = FALSE)

plotBCV(y_female,
        xlab = "Average log CPM",
        ylab = "Biological coefficient of variation - female",
        pch = 16,
        cex = 0.2,
        col.common = "red",
        col.trend = "blue",
        col.tagwise = "black")

plot(bcv_table_male$BCV, bcv_table_female$BCV)

# Density plot of the BCVs
# Add a column to each data frame to indicate the sex
bcv_table_male <- bcv_table_male %>%
  mutate(sex = "male")

bcv_table_female <- bcv_table_female %>%
  mutate(sex = "female")

# Combine the data frames
bcv_combined <- bind_rows(bcv_table_male, bcv_table_female)

# Create the density plot
ggplot(bcv_combined, aes(x = BCV,
                         fill = sex,
                         colour = sex)) +
  geom_density(alpha = 0.5) +
  labs(title = "Density Plot of BCV by Fetal Sex",
       x = "BCV",
       y = "Density") +
  theme_minimal()

```

# Estimate BCV (all samples together)
```{r}
dge_list_obj <- estimateCommonDisp(dge_list_obj,
                                   verbose=TRUE)
dge_list_obj <- estimateTrendedDisp(dge_list_obj)
dge_list_obj <- estimateTagwiseDisp(dge_list_obj)
# plot the BCV vs log2CPM
plotBCV(dge_list_obj)

# Extract gene-wise BCV values
gene_wise_bcv <- sqrt(dge_list_obj$tagwise.dispersion)

# Create a data frame with gene names and BCV values
bcv_table <- data.frame(
  Gene = rownames(dge_list_obj),
  BCV = gene_wise_bcv,
  AveExpr = dge_list_obj$AveLogCPM
)

# Sort the table by BCV (descending order)
bcv_table_sorted <- bcv_table[order(-bcv_table$BCV), ]

# View the top rows of the table
head(bcv_table_sorted)

# save the table to a CSV file:
# write.csv(bcv_table_sorted, file = "gene_wise_bcv_results.csv", row.names = FALSE)
```

# Calculate log2CPM and save table as .csv

```{r}
log2CPM <- edgeR::cpm(dge_list_obj, log = TRUE)

# edgeR::cpm(dge_list_obj, log = TRUE) %>%
#   as.data.frame() %>%
#   tibble::rownames_to_column("ensembl") %>%
#   write_csv(., file = output_log2CPM_file,
#            col_names = TRUE)

```

# Differential Expression
- covariates are either  
  - maternal age  
  - maternal BMI  
```{r}
## No covariates
#Create and write out design matrix
model_noCovariate <- '~0 + sex_outcome'
# establish the design matrix
design_sex_PE <- model.matrix(formula(model_noCovariate),
                              data = dge_list_obj$samples)
# make the column names a little nicer
colnames(design_sex_PE) <- c(sex_outcome_levels)

# # save a csv of the design matrix
# write_csv(data.frame(design_sex_PE) %>%
#             tibble::rownames_to_column("samplename"),
#           file = outDesignMatrix_file,
#           col_names = TRUE)
```

```{r}
# perform the voom
voom_sex_PE <- voom(dge_list_obj, design_sex_PE, plot = TRUE)

# fit the linear model
fit_sex_PE <- lmFit(voom_sex_PE, design_sex_PE)
```

## allTable and residuals for n=58 

```{r}
# perform bayesian adjustment
fit_sex_PE <- eBayes(fit_sex_PE)

# check the coefficients
full_model_coefficients <- fit_sex_PE$coefficients
saveRDS(full_model_coefficients,
        file = file.path(outdir, "full_model_coefficients.rds"))
dim(full_model_coefficients)

# summary table of the sex PE fit
summary(decideTests(fit_sex_PE,
                    adjust.method = "fdr",
                    p.value = 0.05))

# all DE results for the female_PE comparison
allTable_PE <- topTable(fit_sex_PE,
                             # coef = "female_PE",
                             n = Inf,
                             sort = "none")

```

## Sex-specific contrast, fit and residuals

```{r}
# set the contrasts - this makes it easier to "see" what we're testing
contrast_sex_PE <- makeContrasts(female_PE = F_PE-F_Control,
                                 male_PE = M_PE-M_Control,
                                 levels = design_sex_PE)
contrast_sex_PE

# fit a linear regression to the contrast questions
contrast_fit_sex_PE <- contrasts.fit(fit_sex_PE, contrast_sex_PE)

# perform bayesian adjustment
contrast_fit_sex_PE <- eBayes(contrast_fit_sex_PE)

# grab the residuals
res_post_dedup <- residuals(fit_sex_PE,voom_sex_PE)
saveRDS(res_post_dedup,
        file = file.path(outdir, "residuals.rds"))

# grab the fitted values
fitted_post_dedup <- fitted(fit_sex_PE)
saveRDS(fitted_post_dedup,
        file = file.path(outdir, "fitted_values.rds"))

# check the coefficients
coefficients <- contrast_fit_sex_PE$coefficients
saveRDS(coefficients,
        file = file.path(outdir, "coefficients.rds"))
dim(contrast_fit_sex_PE$coefficients)
```

```{r}
# summary table of the sex PE fit
summary(decideTests(contrast_fit_sex_PE,
                    adjust.method = "fdr",
                    p.value = 0.05))

# all DE results for the female_PE comparison
allTable_female_PE <- topTable(contrast_fit_sex_PE,
                             coef = "female_PE",
                             n = Inf,
                             sort = "p")
# create output files and save
write_csv(allTable_female_PE,
          file=file.path(outdir, "allTable_female.csv"))

# write_csv(allTable_female_PE,
#           file="/media/sf_D_DRIVE/VM_Projects/20240723_compare_pre_post_dedup/allTable_female_post_dedup.csv")


# all DE results for the male comparison
allTable_male_PE <- topTable(contrast_fit_sex_PE,
                             coef = "male_PE",
                             n = Inf,
                             sort = "p")
write_csv(allTable_male_PE,
          file=file.path(outdir, "allTable_male.csv"))

# write_csv(allTable_male_PE,
#           file="/media/sf_D_DRIVE/VM_Projects/20240723_compare_pre_post_dedup/allTable_male_post_dedup.csv")

topTable_male_pe <- topTable(contrast_fit_sex_PE,
                             coef = "male_PE",
                             n = Inf,
                             sort = "p",
                             p = 0.05,
                             adjust.method = "fdr")

# create output files and save
write_csv(topTable_male_pe,
          file="/home/smit1924/preeclampsia_sex_chromosome_informed/noCovariate_output/topTable_male_pe.csv")

# write out a few summary tables

# Count occurrences for both conditions in a single pipeline
seqname_counts <- allTable_male %>%
  dplyr::filter(., adj.P.Val < 0.05) %>%
  mutate(condition = ifelse(logFC < 0, "down", "up")) %>%
  group_by(seqnames, condition) %>%
  summarise(count = n(), .groups = "drop")

# View the result
print(seqname_counts)

df <- left_join(allTable_female_PE %>% dplyr::select(-t, -P.Value, -B),
               allTable_male_PE %>% dplyr::select(-t, -P.Value, -B),
               by=c('hgnc_symbol', 'ensembl'),
          suffix=c("_F", "_M"))

coeff_and_sigma_df <- cbind(fit_sex_PE$coefficients, sigma=fit_sex_PE$sigma) %>%
  as.data.frame() %>%
  rownames_to_column(var="ensembl")

df <- left_join(coeff_and_sigma_df,
               df,
               by="ensembl")

write_csv(df %>% relocate(hgnc_symbol, .after=ensembl),
          file=output_spreadsheet_file)

df2 <- left_join(allTable_female_PE %>% dplyr::select(ensembl, hgnc_symbol, t, P.Value, adj.P.Val),
               allTable_male_PE %>% dplyr::select(ensembl, hgnc_symbol, t, P.Value, adj.P.Val),
               by=c('hgnc_symbol', 'ensembl'),
          suffix=c("_F", "_M"))

write_csv(df2 %>% relocate(hgnc_symbol, .after=ensembl),
          file=output_pValue_file)

```

## Sanity plot of DE genes

```{r}
# let's take a look at the DE genes with the highest logFC that also have a "decent" expression level

# row center the counts
# do this by subtracting the row(gene) mean from each count

# center with 'scale()'
# setting `scale = FALSE` subtracts the mean rather than zero centering
center_scale <- function(x) {
    scale(x, scale = FALSE)
}

# apply it
row_centered_counts <- center_scale(t(dge_list_obj$counts)) %>% 
  t()

# which gene would we like to plot??
gene_to_plot <- "ENSG00000047457"

# png(filename=file.path(outdir, "expression_DE_gene_CP_sexOutcome.png'), width=800, height=800)
# Gene of Interest (GOI)
GOI_PE_UP <- data.frame(ensbl=c("ENSG00000178726", "ENSG00000115602",
                             "ENSG00000174697", "ENSG00000070404",
                             "ENSG00000169495", "ENSG00000136002",
                             "ENSG00000012223", "ENSG00000142623",
                             "ENSG00000224658", "ENSG00000102755",
                             "ENSG00000118849", "ENSG00000165810",
                             "ENSG00000211448", "ENSG00000243836",
                             "ENSG00000150722", "ENSG00000181458",
                             "ENSG00000159399"),
                  gene=c("THBD", "IL1RL1", "LEP", "FSTL3",
                         "HTRA4", "ARHGEF4", "LTF", "PADI1",
                         "RP11-631F7.1", "FLT1", "RARRES1",
                         "BTNL9", "DIO2", "WDR86-AS1", "PPP1R1C",
                         "TMEM45A", "HK2"),
                  direction = "up")

GOI_PE_male_UP <- topTable_male_pe %>%
  arrange(desc(logFC)) %>%
  dplyr::select(., ensbl = ensembl_gene_id, gene = hgnc_symbol)

GOI_PE_DOWN <- data.frame(ensbl=c("ENSG00000233913", "ENSG00000134548",
                                  "ENSG00000203859", "ENSG00000213088",
                                  "ENSG00000114473", "ENSG00000260719",
                                  "ENSG00000091137", "ENSG00000157064"),
                  gene=c("RPL10P9", "SPX", "HSD3B2", "ACKR1",
                         "IQCG", "AC009133.17", "SLC26A4", "NMNAT2"),
                  direction = "down")

GOI_UP_DOWN <- rbind(GOI_PE_UP, GOI_PE_DOWN)

samples <- dge_list_obj$samples

counts <- edgeR::cpm(dge_list_obj, log = TRUE)

pd <- position_dodge(width = 0.5)

lcpm <- edgeR::cpm(dge_list_obj, log = TRUE)

## Print all of the plots at once
for (i in 1:nrow(GOI_PE_UP)) {
melt_lcpm <- melt(subset(lcpm, rownames(lcpm) %in% GOI_PE_UP$ensbl[i])) %>% 
  as.data.frame() %>% 
  set_colnames(c("ensbl", "samplename", "log2cpm")) %>% 
  # tibble::rownames_to_column("samplename") %>% 
  left_join(., dge_list_obj$samples[, c("samplename", "Outcome", "Sex")], by = "samplename") 
  
beeswarm <- ggplot(melt_lcpm, aes(x = Sex,
                                y = log2cpm,
                                fill = Outcome)) +
  stat_boxplot(geom="errorbar",
               position = pd,
               width = 0.2) +
  geom_boxplot(width = 0.5,
               position = pd) +
  #     geom_boxplot(aes(x = Fetal.Sex,
  #                  y =  log2cpm,
  #                  fill = simpleOutcome)) +
  stat_summary(fun = mean,
               geom = "point",
               size=2,
               shape = 5,
               position = position_dodge(0.5)) +
  geom_beeswarm(dodge.width = 0.6, cex = 2) +
  theme_bw(base_size = 18) +
  ylab(paste("log2 CPM Counts", GOI_PE_UP$ensbl[i], GOI_PE_UP$gene[i], sep = " - ")) +
  theme(axis.text.x = element_text(colour = "black")) +
  theme(legend.position = "bottom")
print(beeswarm)
}

# graphics.off()
```

## Volcano plot for visualisation of differential expression
### male

```{r volcano plots male}

decide <- decideTests(contrast_fit_sex_PE)
d2 <- topTable(contrast_fit_sex_PE, coef = 2, n = Inf, sort = "p") #[,c(2,6)]
d2$threshold <- 0
d2$threshold[0:nrow(filter(allTable_male_PE, adj.P.Val < 0.05))] <- 1
d2$threshold <- as.factor(d2$threshold)
d3 <- decideTests(contrast_fit_sex_PE) %>%
  as.data.frame()
d3 <- d3[2]
d4 <- left_join(tibble::rownames_to_column(d2), (tibble::rownames_to_column(d3)),
                       by = "rowname")
d4$male_PE <- as.factor(d4$male_PE)
d4$neg.log10FDR <- -log10(d4$adj.P.Val)
# set colours for the volcano plot
colour <- c("red", "black", "darkgreen")
FDR_FC_Decide <- decideTests(contrast_fit_sex_PE, lfc = 0)[,"male_PE"] %>%
  as.data.frame() %>%
  set_colnames("FDR_FC_Decide") %>%
  tibble::rownames_to_column() %>%
  left_join(., d4, by = "rowname")
FDR_FC_Decide$FDR_FC_Decide <- as.factor(FDR_FC_Decide$FDR_FC_Decide)

volcano_male <- ggplot(data = FDR_FC_Decide,
                  aes(x = logFC,
                      y = neg.log10FDR,
                      label = rowname,
                      colour = FDR_FC_Decide)) +
  geom_point(alpha=0.9, size=2.00) +
  xlab(expression(Log[2]*" Fold Change")) + ylab(expression("-ve "*Log[10]*" FDR")) +
  scale_color_manual(values = colour, name="mRNA\nRegulation",
                         breaks = c("-1", "0", "1"),
                         labels = c("Down-regulated", "Not Significant", "Up-regulated")) +
  theme(plot.title = element_text(hjust = 0.5)) +
  geom_hline(yintercept = -log10(.05), linetype = "dotted") +
  geom_vline(xintercept = c(-1, 1), linetype = "dotted") +
  theme_bw() +
  theme(text = element_text(size=18)) +
  theme(axis.text.x = element_text(colour = "black",
                                   face = "bold"),
        axis.text.y = element_text(colour = "black",
                                   face = "bold")) +
  theme(axis.title.x = element_text(colour = "black",
                                   face = "bold"),
        axis.title.y = element_text(colour = "black",
                                   face = "bold")) +
  geom_label_repel(data = dplyr::filter(FDR_FC_Decide, FDR_FC_Decide==1),
                  aes(label = hgnc_symbol),
                  fill = "white",
                  xlim = c(7, 0),
                  max.overlaps = Inf,
                  show.legend = FALSE) +
  geom_label_repel(data = dplyr::filter(FDR_FC_Decide, FDR_FC_Decide==-1),
                  aes(label = hgnc_symbol),
                  fill = "white",
                  xlim = c(-7, 0),
                  max.overlaps = Inf,
                  show.legend = FALSE) +
  theme(legend.position="bottom")

ggsave(plot = volcano_male,
       filename = file.path(projectDir, "deduplicated_clearBox/output/volcano_male.png"),
       units = "in",
       width = 12,
       height = 12,
       dpi = 300)


```

### female

```{r volcano plots female}

decide <- decideTests(contrast_fit_sex_PE)
d2 <- topTable(contrast_fit_sex_PE, coef = 1, n = Inf, sort = "p") #[,c(2,6)]
d2$threshold <- 0
d2$threshold[0:nrow(filter(allTable_female_PE, adj.P.Val < 0.05))] <- 1
d2$threshold <- as.factor(d2$threshold)
d3 <- decideTests(contrast_fit_sex_PE) %>%
  as.data.frame()
d3 <- d3[1]
d4 <- left_join(tibble::rownames_to_column(d2), (tibble::rownames_to_column(d3)),
                       by = "rowname")
d4$female_PE <- as.factor(d4$female_PE)
d4$neg.log10FDR <- -log10(d4$adj.P.Val)
# set colours for the volcano plot
colour <- c("red", "black", "darkgreen")
FDR_FC_Decide <- decideTests(contrast_fit_sex_PE, lfc = 0)[,"female_PE"] %>%
  as.data.frame() %>%
  set_colnames("FDR_FC_Decide") %>%
  tibble::rownames_to_column() %>%
  left_join(., d4, by = "rowname")
FDR_FC_Decide$FDR_FC_Decide <- as.factor(FDR_FC_Decide$FDR_FC_Decide)

volcano_female <- ggplot(data = FDR_FC_Decide,
                  aes(x = logFC,
                      y = neg.log10FDR,
                      label = rowname,
                      colour = FDR_FC_Decide)) +
  geom_point(alpha=0.9, size=2.00) +
  xlab(expression(Log[2]*" Fold Change")) + ylab(expression("-ve "*Log[10]*" FDR")) +
  scale_color_manual(values = colour, name="mRNA\nRegulation",
                         breaks = c("-1", "0", "1"),
                         labels = c("Down-regulated", "Not Significant", "Up-regulated")) +
  theme(plot.title = element_text(hjust = 0.5)) +
  geom_hline(yintercept = -log10(.05), linetype = "dotted") +
  geom_vline(xintercept = c(-1, 1), linetype = "dotted") +
  theme_bw() +
  theme(text = element_text(size=18)) +
  theme(axis.text.x = element_text(colour = "black",
                                   face = "bold"),
        axis.text.y = element_text(colour = "black",
                                   face = "bold")) +
  theme(axis.title.x = element_text(colour = "black",
                                   face = "bold"),
        axis.title.y = element_text(colour = "black",
                                   face = "bold")) +
  geom_label_repel(data = dplyr::filter(FDR_FC_Decide, FDR_FC_Decide==1),
                  aes(label = hgnc_symbol),
                  fill = "white",
                  xlim = c(7, 0),
                  max.overlaps = Inf,
                  show.legend = FALSE) +
  geom_label_repel(data = dplyr::filter(FDR_FC_Decide, FDR_FC_Decide==-1),
                  aes(label = hgnc_symbol),
                  fill = "white",
                  xlim = c(-7, 0),
                  max.overlaps = Inf,
                  show.legend = FALSE) +
  theme(legend.position="bottom")

ggsave(plot = volcano_female,
       filename = file.path(projectDir, "noCovariate_output/volcano_female.pdf"),
       units = "in",
       width = 12,
       height = 12,
       dpi = 300)


```

# Heatmaps
```{r}
# Modify ordering of the clusters using clustering callback option
callback = function(hc, mat){
    sv = svd(t(mat))$v[,1]
    dend = reorder(as.dendrogram(hc), wts = sv)
    as.hclust(dend)
}

males <- dplyr::filter(dge_list_obj$samples, Sex == "M")

annotation_cols <- metadata[, c("samplename", "Sex", "Outcome")] %>%
  tibble::column_to_rownames("samplename") %>%
  dplyr::filter(., Sex == "M")

annotation_rows <- topTable_male_pe %>%
  dplyr::mutate(foldChange = dplyr::case_when(
    logFC > 0 ~ "up",
    logFC < 0 ~ "down"
  )) %>%
  dplyr::select(., foldChange)
# Define the color palette for row annotations
annotation_colors <- list(
  foldChange = c("down" = "red", "up" = "green")
)

scaled_exp <- scale(t(subset(log2CPM, rownames(log2CPM) %in% dplyr::filter(FDR_FC_Decide, threshold == 1)$ensembl)))
scaled_male <- subset(scaled_exp, rownames(scaled_exp) %in% males$samplename)

png(file=file.path(outdir, "heatmap_scaled_sig_AdjPVal_M.png"),
    width=12,
    height=8,
    units = "in",
    res = 150)
pheatmap(t(scaled_male),
         annotation_col = annotation_cols,
         annotation_row = annotation_rows,
         annotation_colors = annotation_colors,
         show_rownames = FALSE,
         show_colnames = TRUE,
         cutree_cols = 3,
         treeheight_row = FALSE,
         clustering_callback = callback)
dev.off()

```

# Session information
```{r session info}

sessionInfo()

```