---
title: "tpm_cpm_calc_allSample_deduplicated"
author: "Melanie Smith"
date: "26 June 2024"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

This script converts the counts generated by `featureCounts` into transcripts per million (TPMs). Conversion to TPMs is required before we can compare the bulk counts with scRNAseq counts. In this case, the comparison will be used for cell-type deconvolution using CIBERSORTx.  

Here we take the raw counts generated in the clearBox analysys and compare them to the de-duplicated counts.  
Only the six samples we already have deduplicated counts for will be compared.  

# Load Libraries

```{r}
library(IRanges)
library(GenomicFeatures)
library(edgeR)
library(readxl)
library(readr)
library(magrittr)
library(stringr)
library(dplyr)

library(ggplot2)
```

# Project directories and input/output files

```{r}
# Set project directory
projectDir <- "/home/smit1924/preeclampsia_sex_chromosome_informed"
outdir <- file.path(projectDir, "deduplicated_clearBox/output")
expt_name <- "deduplicated_bams"

# input files
## Genome files
input_genome_gtf_file <- file.path(projectDir, "clearBoxCleanData/gencode.v29.annotation.gtf")
input_genome_gff3_file <- file.path(projectDir, "clearBoxCleanData/gencode.v29.chr_patch_hapl_scaff.annotation.gff3")

## Raw count files
## read counts generated from deduplicated bams (these were annotated in sex informed STAR)
input_counts_deduplicated_file <- file.path(projectDir, "deduplicated_clearBox/readCounts/deduplicated_s2_readCounts.txt")

## Output TPM file
## This file will contain tpms for 6 standard and 6 deduplicated bams (paired)
output_tpm_allSample_deduplicated_bulkCountMatrix_file <- file.path(outdir, "tpm_allSample_deduplicated_bulkCountMatrix.txt")

## Output CPM file
## This file will contain cpms for 6 standard and 6 deduplicated bams (paired)
output_cpm_allSample_deduplicated_bulkCountMatrix_file <- file.path(outdir, "cpm_allSample_deduplicated_bulkCountMatrix.txt")

# set filtering criteria
filterCPM <- 1
numSamples <- 5

## set max digits
options(digits = 3)
```

# Import files
## GTF file
Before I can transform my raw counts to TPMs I need to calculate the gene lengths from the original .gtf file.   

```{r}

# Try it
# Specifics of calculating rpkm: https://support.bioconductor.org/p/64585/

# 1. Get gene lengths
# https://www.biostars.org/p/83901/
# Or using GTFtools, but keep chromosome limitations in mind:
# http://genomespot.blogspot.com/2019/01/using-gtf-tools-to-get-gene-lengths.html
# GTFtools offers flexibility about how to calculate the gene length (median, mean, etc.)

# Note, this is the gtf used in the RNA-Seq pipeline to annotate hg38
txdb <- GenomicFeatures::makeTxDbFromGFF(file = file.path(input_genome_gtf_file), format = "gtf")
# Collect exons per gene id
exon_list_per_gene <- GenomicFeatures::exonsBy(txdb, by = "gene")
exon_list_per_gene

```

## Calculate exonic gene sizes

```{r}

# For each gene, reduce all exons to a set of non-overlapping exons, calculate their length (widths) and sum them
exonic_gene_sizes <- 
  sum(IRanges::width(IRanges::reduce(exon_list_per_gene))) %>% 
  tibble::as_tibble(rownames = NA) %>% 
  tibble::rownames_to_column(var = "gene") %>%
  dplyr::select(., gene, length = value)
exonic_gene_sizes

# calculate the number of unique genes in the "gene" column of the exonic_gene_sizes data frame.
dim(exonic_gene_sizes) #58721 2
dim(exonic_gene_sizes[!duplicated(exonic_gene_sizes$gene), ]) #58721 2
# at this point there are no duplicated ensembl ids

```

## Import the deduplicated count file

```{r import gene counts - deduplicated}

# import the counts table
# counts are annotated to GRCh38
rawCounts_deduplicated <- read.delim(file = input_counts_deduplicated_file) %>%
  as.data.frame()
# tidy up the column names
colnames(rawCounts_deduplicated) <- gsub("X.media.sf_D_DRIVE.VM_Projects.20240619_deduplicateBams.deduplicated_data.|_marked_duplicates.bam|_T.*|_",
                                 "",
                                 colnames(rawCounts_deduplicated))

# remove the genes from the PAR_Y (all have zero counts) they look like "ENSG00000185203.12_PAR_Y"
rawCounts_deduplicated %<>% dplyr::filter(., !grepl("PAR_Y",Geneid))
dim(rawCounts_deduplicated) #58686 61
# append "_dedup"" to the column name so we don't have any duplicated sample IDs
rawCounts_deduplicated %<>% rename_at(vars(starts_with("SC") | starts_with("ST")), ~paste0(., "_dedup"))
# calculate the number of unique ensembl ids
dim(rawCounts_deduplicated[!duplicated(rawCounts_deduplicated$Geneid), ]) #58676 61
# at this point there are no duplicated ensembl ids

```

## Calculate TPMs

```{r}

# 2. Process raw count data in edgeR again, as per melanieRawCountsProcessing_v2.0.3.rmd
# Add gene length to expression data
raw_data <- rawCounts_deduplicated %>%
  dplyr::mutate(Length = exonic_gene_sizes$length[match(Geneid, exonic_gene_sizes$gene)]) %>%
  # move the ensemble ids into the rownames
  tibble::column_to_rownames("Geneid")
# check in on the dimensions
head(raw_data)
dim(raw_data) # 58676 61

# establish the DGEList object
y <- edgeR::DGEList(counts = raw_data %>% dplyr::select(-Length),
                    genes = raw_data %>% dplyr::select(Length))

y
y$genes # A Length column is present 
dim(y) # 58676 60

# Identify genes expressed at > 1 cpm in at least 5 sample
keep <- rowSums(cpm(y) > filterCPM) >= numSamples
table(keep) # Keeping 17345 genes

# Apply filter
y <- y[keep, , keep.lib.sizes = FALSE]
dim(y) # 17345  60

head(y$counts) # Gene names are set as rownames (still have the ENSG0000000000.0 versions)
y$samples$lib.size # 16913758  8163430 17285116 10074871 13313364  7225362  6622031  5721265 12790534  6590531  8475766  3603947  8224849  4154035  5454318  6768880 13326237 6601585  6326135  7868409  8340918  4906404 11505664  8044491 12963902 13341437 12301644  5699083 12099281  6917674 12614894 12633994 12233938 15563744  15697737  4281081  6068986  8440731 10067135 14360867 14097864 14130817  6518697  9801805 15941999 11854683 15456135 17130922  7130448  7188635 13709741 12469663  7460158  5886234  7062633  6341479  8235146  8832833 13715280  6614832

# TMM normalisation 
y$samples$norm.factors # currently all normalisation factors == 1
y <- calcNormFactors(y, method = "TMM")
y$samples$norm.factors # 1.089 0.948 1.007 1.177 0.999 0.978 1.066 1.067 1.199 1.168 0.956 0.953 0.920 1.002 1.069 0.999 0.963 1.057 1.026 0.990 0.900 1.165 1.024 0.969 1.072 1.101 0.948 0.958 1.091 0.927 0.962 1.048 1.044 0.936 1.049 1.167 0.888 0.999 0.945 0.994 0.959 0.969 1.170 1.206 0.945 0.950 0.998 0.959 0.908 0.941 0.984 0.924 0.889 1.092 0.892 0.917 0.901 0.903 0.934 0.945

# Now calculate RPKM: https://support.bioconductor.org/p/64585/
rpkm <- rpkm(y)

# And TPM: https://www.biostars.org/p/388584/#394579
tpm <- t( t(rpkm) / colSums(rpkm) ) * 1e6
# These TPM values now incorporate the TMM factors

# generate a table of cpms
cpm <- edgeR::cpm(y, log = FALSE)

# save a copy of the cpms for Kat
cpm %>%
  write.table(file = output_cpm_allSample_deduplicated_bulkCountMatrix_file,
              quote = FALSE,
              sep = "\t",
              col.names = TRUE,
              row.names = TRUE)

```
# Plot the raw library sizes

```{r}
y$samples %>%
  tibble::rownames_to_column("samplename") %>%
  mutate(color = case_when(
    lib.size < 5000000 ~ "red",
    lib.size >= 5000000 & lib.size < 10000000 ~ "blue",
    lib.size >= 10000000 ~ "darkgreen"
  )) %>%
  ggplot(aes(x = reorder(samplename, -lib.size), y = lib.size, fill = color)) +
  geom_bar(stat = "identity") +
  geom_hline(yintercept = 5000000, linetype = "dashed", color = "red") +
  geom_hline(yintercept = 10000000, linetype = "dashed", color = "darkgreen") +
  labs(y = "Library size (total number of mapped and quantified reads)",
       x = "Samplename") +
  scale_fill_identity() +
  coord_flip()


temp <- y$samples %>%
  tibble::rownames_to_column("samplename")
```

## lets add the gene name information
### Import gencode V29 gff3 file

```{r}

# import the gencodev32_gff_inFile file
all_gencode_v29 <- rtracklayer::import(input_genome_gff3_file)
# this file contains more information than we need here
# subset out only the columns we need
gene_data <- data.frame(ensembl_gene_id = all_gencode_v29@elementMetadata$gene_id,
                        hgnc_symbol = all_gencode_v29@elementMetadata$gene_name,
                        seqnames = all_gencode_v29@seqnames)
# we're left with multiple identical rows, keep one each
gencode_v29_gene_id_symbol_chr_biotype <- gene_data %>%
  dplyr::distinct(, .keep_all = TRUE)
# which leaves us with the following dimensions where there is a unique combination of ensembl ID, hgnc symbol and chromosome
dim(gencode_v29_gene_id_symbol_chr_biotype) # 64837 3
# this does however, leave us with duplicate ensembl IDs
# 64792 3
# after generating a table to see the duplicates I can see that this is an X/Y chromosome issue. All genes are assigned to both chromosomes
duplicate_rows <- gencode_v29_gene_id_symbol_chr_biotype[duplicated(gencode_v29_gene_id_symbol_chr_biotype$ensembl_gene_id) | duplicated(gencode_v29_gene_id_symbol_chr_biotype$ensembl_gene_id, fromLast = TRUE),]
# remove duplicates (we don't use the chromosome information for anything)
gencode_v29_gene_id_symbol_chr_biotype <- gencode_v29_gene_id_symbol_chr_biotype[!duplicated(gencode_v29_gene_id_symbol_chr_biotype$ensembl_gene_id), ]
```

### add the gene info to the tpms

```{r}
# join the counts with the gene info
tpm_gene_id_symbol_chr_biotype <- data.frame(tpm) %>%
  tibble::rownames_to_column("ensembl_gene_id") %>%
  dplyr::left_join(., gencode_v29_gene_id_symbol_chr_biotype, by = "ensembl_gene_id") %>%
  dplyr::select(colnames(gencode_v29_gene_id_symbol_chr_biotype), colnames(data.frame(tpm)))

dim(tpm_gene_id_symbol_chr_biotype) # 17345 63

# save a copy of the tpm counts with only the hgnc gene symbol
tpm_gene_id_symbol_chr_biotype %>%
  # remove the ensembl gene id and the chromosome name
  dplyr::select(., -ensembl_gene_id, -seqnames) %>%
  # save TPM count file for CIBERSTORTx
  write.table(file = output_tpm_allSample_deduplicated_bulkCountMatrix_file,
              quote = FALSE,
              sep = "\t",
              col.names = TRUE,
              row.names = FALSE)

```

## Session information

```{r session info}

sessionInfo()

```