---
title: "tpm_cpm_calc_sixSample"
author: "Melanie Smith"
date: "1 May 2024"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

This script converts the counts generated by `featureCounts` into transcripts per million (TPMs). Conversion to TPMs is required before we can compare the bulk counts with scRNAseq counts. In this case, the comparison will be used for cell-type deconvolution using CIBERSORTx.  

Here we take the raw counts generated in the clearBox analysys and compare them to the de-duplicated counts.  
Only the six samples we already have deduplicated counts for will be compared.  

# Load Libraries

```{r}
library(IRanges)
library(GenomicFeatures)
library(edgeR)
library(readxl)
library(readr)
library(magrittr)
library(stringr)
library(dplyr)
```

# Project directories and input/output files

```{r}
# Set project directory
projectDir <- "/home/smit1924/preeclampsia_sex_chromosome_informed"
outdir <- file.path(projectDir, "deduplicated_clearBox/output")
expt_name <- "deduplicated_bams"

# input files
## Genome files
input_genome_gtf_file <- file.path(projectDir, "clearBoxCleanData/gencode.v29.annotation.gtf")
input_genome_gff3_file <- file.path(projectDir, "clearBoxCleanData/gencode.v29.chr_patch_hapl_scaff.annotation.gff3")

## Raw count files
## read counts generated from deduplicated bams (these were annotated in sex informed STAR)
input_counts_deduplicated_file <- file.path(projectDir, "deduplicated_clearBox/readCounts/20240429_deduplicated_s2_readCounts.txt")
## read counts generated from standard bams
input_counts_female_file <- file.path(projectDir, "clearBoxRawData/featureCounts_v2.0.3_female_readCounts.txt")
input_counts_male_file <- file.path(projectDir, "clearBoxRawData/featureCounts_v2.0.3_male_readCounts.txt")

## CPM matrix from the full, pre-deduplicated counts (use the genes from this matrix to subset the deduplicated matrix)
## The counts were generated previously from the standard bam read counts and contain all female and male samples
input_20230608_log2cpms_no_gene_names_file <- file.path(projectDir, "clearBoxCleanData/20230608_log2cpms_no_gene_names.txt")

## Output TPM file
## This file will contain tpms for 6 standard and 6 deduplicated bams (paired)
output_tpm_12_sample_combined_bulkCountMatrix_file <- file.path(outdir, "tpm_12_sample_combined_bulkCountMatrix.txt")

## Output CPM file
## This file will contain cpms for 6 standard and 6 deduplicated bams (paired)
output_cpm_12_sample_combined_bulkCountMatrix_file <- file.path(outdir, "cpm_12_sample_combined_bulkCountMatrix.txt")

## deduplicated samplenames
## These are the only samples that have been deduplicated up to now
dedup_samples <- c("STP0148", "SCP3954", "STP1206", "STP0932", "SCP4733", "STP0596")

## set max digits
options(digits = 3)
```

# Import files
## GTF file
Before I can transform my raw counts to TPMs I need to calculate the gene lengths from the original .gtf file.   

```{r}

# Try it
# Specifics of calculating rpkm: https://support.bioconductor.org/p/64585/

# 1. Get gene lengths
# https://www.biostars.org/p/83901/
# Or using GTFtools, but keep chromosome limitations in mind:
# http://genomespot.blogspot.com/2019/01/using-gtf-tools-to-get-gene-lengths.html
# GTFtools offers flexibility about how to calculate the gene length (median, mean, etc.)

# Note, this is the gtf used in the RNA-Seq pipeline to annotate hg38
txdb <- GenomicFeatures::makeTxDbFromGFF(file = file.path(input_genome_gtf_file), format = "gtf")
# Collect exons per gene id
exon_list_per_gene <- GenomicFeatures::exonsBy(txdb, by = "gene")
exon_list_per_gene

```

## Calculate exonic gene sizes

```{r}

# For each gene, reduce all exons to a set of non-overlapping exons, calculate their length (widths) and sum them
exonic_gene_sizes <- 
  sum(IRanges::width(IRanges::reduce(exon_list_per_gene))) %>% 
  tibble::as_tibble(rownames = NA) %>% 
  tibble::rownames_to_column(var = "gene") %>%
  dplyr::select(., gene, length = value)
exonic_gene_sizes

# calculate the number of unique genes in the "gene" column of the exonic_gene_sizes data frame.
dim(exonic_gene_sizes) #58721 2
dim(exonic_gene_sizes[!duplicated(exonic_gene_sizes$gene), ]) #58721 2
# at this point there are no duplicated ensembl ids

```

## Import the deduplicated count file

```{r import gene counts - deduplicated}

# import the counts table
# counts are annotated to GRCh38
rawCounts_deduplicated <- read.delim(file = input_counts_deduplicated_file) %>%
  as.data.frame()
# tidy up the column names
colnames(rawCounts_deduplicated) <- gsub("X.scratch.user.smit1924.preeclampsia_sex_chromosome_informed.deduplicated_clearBox.deduplicated_bams.|_marked_duplicates.bam",
                                 "",
                                 colnames(rawCounts_deduplicated))

# remove the genes from the PAR_Y (all have zero counts) they look like "ENSG00000185203.12_PAR_Y"
rawCounts_deduplicated %<>% dplyr::filter(., !grepl("PAR_Y",Geneid))
dim(rawCounts_deduplicated) #58686 7
# append "_dedup"" to the column name so we don't have any duplicated sample IDs
rawCounts_deduplicated %<>% rename_at(vars(starts_with("SC") | starts_with("ST")), ~paste0(., "_dedup"))
# calculate the number of unique ensembl ids
dim(rawCounts_deduplicated[!duplicated(rawCounts_deduplicated$Geneid), ]) #58676 7
# at this point there are no duplicated ensembl ids


```

## Import the Female count file

```{r import gene counts - deduplicated}

# import the counts table
# counts are annotated to GRCh38
rawCounts_female <- read.delim(file = input_counts_female_file) %>%
  as.data.frame()
# tidy up the column names
colnames(rawCounts_female) <- gsub("X.scratch.user.smit1924.20230608_female_grch38.aligned_data.|_T.*|_",
                                 "",
                                 colnames(rawCounts_female))
# remove the genes from the PAR_Y (all have zero counts) they look like "ENSG00000185203.12_PAR_Y"
rawCounts_female %<>% dplyr::filter(., !grepl("PAR_Y",Geneid))
dim(rawCounts_female) #58686 39
# calculate the number of unique ensembl ids
dim(rawCounts_female[!duplicated(rawCounts_female$Geneid), ]) #58676 39
# at this point there are no duplicated ensembl ids

```

## Import the Male count file

```{r import gene counts - deduplicated}

# import the counts table
# counts are annotated to GRCh38
rawCounts_male <- read.delim(file = input_counts_male_file) %>%
  as.data.frame()
# tidy up the column names
colnames(rawCounts_male) <- gsub("X.scratch.user.smit1924.20230608_male_grch38.aligned_data.|_T.*|_",
                                 "",
                                 colnames(rawCounts_male))
# remove the genes from the PAR_Y (all have zero counts) they look like "ENSG00000185203.12_PAR_Y"
rawCounts_male %<>% dplyr::filter(., !grepl("PAR_Y",Geneid))
dim(rawCounts_male) #58686 24
# calculate the number of unique ensembl ids
dim(rawCounts_male[!duplicated(rawCounts_male$Geneid), ]) #58676 24
# at this point there are no duplicated ensembl ids

```

## import input_20230608_log2cpms_no_gene_names_file
- This matrix will be used to subset the combined tpms to ensure we have the same gene set as previous cibersortx runs  
- Subset the tpms using the ensembl id with version number intact

```{r import cpms - not deduplicated}
# import the tpms table
log2cpms_no_gene_names <- read.delim(file = input_20230608_log2cpms_no_gene_names_file) %>%
  as.data.frame()
dim(log2cpms_no_gene_names) #13767 62
```

## Combine counts

```{r}
## test to ensure the rownames are identical
identical(rownames(rawCounts_female),
                   rownames(rawCounts_male))
identical(rownames(rawCounts_female),
                   rownames(rawCounts_deduplicated))

# combined the male and female counts
combined_rawCounts <- dplyr::left_join(rawCounts_female,
                                       rawCounts_male,
                                       by = "Geneid") %>%
  # subset to only include the six samples that pair with the deduplicated bams
  dplyr::select(., Geneid, dedup_samples) %>%
  # join the deduplicated counts
  dplyr::full_join(., rawCounts_deduplicated,
                   by = "Geneid")
dim(combined_rawCounts) # 58676 13

```

## Calculate TPMs

```{r}

# 2. Process raw count data in edgeR again, as per melanieRawCountsProcessing_v2.0.3.rmd
# Add gene length to expression data
raw_data <- combined_rawCounts %>%
  dplyr::mutate(Length = exonic_gene_sizes$length[match(Geneid, exonic_gene_sizes$gene)]) %>%
  # move the ensemble ids into the rownames
  tibble::column_to_rownames("Geneid")
# check in on the dimensions
head(raw_data)
dim(raw_data) # 58676 13

# establish the DGEList object
y <- edgeR::DGEList(counts = raw_data %>% dplyr::select(-Length),
                    genes = raw_data %>% dplyr::select(Length))

y
y$genes # A Length column is present 
dim(y) # 58676 12

head(y$counts) # Gene names are set as rownames (still have the ENSG0000000000.0 versions)
y$samples$lib.size # 31442737 25759561 11656794 11513311 10617410 12135572 12821534  5711772 15733226  6530654  7074417  6624113

# TMM normalisation 
y$samples$norm.factors # currently all normalisation factors == 1
y <- calcNormFactors(y, method = "TMM")
y$samples$norm.factors # 0.675 0.788 0.735 0.771 0.717 0.851 1.487 1.278 1.270 1.496 1.198 1.258

# Now calculate RPKM: https://support.bioconductor.org/p/64585/
rpkm <- rpkm(y)

# And TPM: https://www.biostars.org/p/388584/#394579
tpm <- t( t(rpkm) / colSums(rpkm) ) * 1e6
# These TPM values now incorporate the TMM factors

# Subset the combined tpms to only contain genes in the pre-deduplicated bulkCountMatrix
idx <- match(rownames(tpm),
             log2cpms_no_gene_names$Geneid)
subsetted_table <- tpm[!is.na(idx),] # 13767 15

# generate a table of cpms
cpm <- edgeR::cpm(y, log = FALSE)
# save a copy of the cpms for Kat
cpm %>%
  write.table(file = output_cpm_12_sample_combined_bulkCountMatrix_file,
              quote = FALSE,
              sep = "\t",
              col.names = TRUE,
              row.names = TRUE)

```

## lets add the gene name information
### Import gencode V29 gff3 file

```{r}

# import the gencodev32_gff_inFile file
all_gencode_v29 <- rtracklayer::import(input_genome_gff3_file)
# this file contains more information than we need here
# subset out only the columns we need
gene_data <- data.frame(ensembl_gene_id = all_gencode_v29@elementMetadata$gene_id,
                        hgnc_symbol = all_gencode_v29@elementMetadata$gene_name,
                        seqnames = all_gencode_v29@seqnames)
# we're left with multiple identical rows, keep one each
gencode_v29_gene_id_symbol_chr_biotype <- gene_data %>%
  dplyr::distinct(, .keep_all = TRUE)
# which leaves us with the following dimensions where there is a unique combination of ensembl ID, hgnc symbol and chromosome
dim(gencode_v29_gene_id_symbol_chr_biotype) # 64837 3
# this does however, leave us with duplicate ensembl IDs
# 64792 3
# after generating a table to see the duplicates I can see that this is an X/Y chromosome issue. All genes are assigned to both chromosomes
duplicate_rows <- gencode_v29_gene_id_symbol_chr_biotype[duplicated(gencode_v29_gene_id_symbol_chr_biotype$ensembl_gene_id) | duplicated(gencode_v29_gene_id_symbol_chr_biotype$ensembl_gene_id, fromLast = TRUE),]
# remove duplicates (we don't use the chromosome information for anything)
gencode_v29_gene_id_symbol_chr_biotype <- gencode_v29_gene_id_symbol_chr_biotype[!duplicated(gencode_v29_gene_id_symbol_chr_biotype$ensembl_gene_id), ]
```

### add the gene info to the tpms

```{r}
# join the counts with the gene info
tpm_gene_id_symbol_chr_biotype <- data.frame(subsetted_table) %>%
  tibble::rownames_to_column("ensembl_gene_id") %>%
  dplyr::left_join(., gencode_v29_gene_id_symbol_chr_biotype, by = "ensembl_gene_id") %>%
  dplyr::select(colnames(gencode_v29_gene_id_symbol_chr_biotype), colnames(data.frame(subsetted_table)))

dim(tpm_gene_id_symbol_chr_biotype) # 13767 15

# save a copy of the tpm counts with only the hgnc gene symbol
tpm_gene_id_symbol_chr_biotype %>%
  # remove the ensembl gene id and the chromosome name
  dplyr::select(., -ensembl_gene_id, -seqnames) %>%
  # save TPM count file for CIBERSTORTx
  write.table(file = output_tpm_12_sample_combined_bulkCountMatrix_file,
              quote = FALSE,
              sep = "\t",
              col.names = TRUE,
              row.names = FALSE)

```
