---
title: "geneLengthCalc"
author: "Melanie Smith"
date: "18 April 2023"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

This script converts the counts generated by `featureCounts` into transcripts per million (TPMs).  
Conversion to TPMs is required before we can compare the bulk counts with scRNAseq counts.  
In this case, the comparison will be used for cell type deconvolution.  

```{r}
library(IRanges)
library(GenomicFeatures)
library(edgeR)
library(readxl)
library(readr)
library(magrittr)

# set project directory
projectDir <- "/home/smit1924/preeclampsia_sex_chromosome_informed"
expt_name <- "deduplicated_bams"

genome_gtf_infile <- file.path(projectDir, "clearBoxCleanData/gencode.v29.annotation.gtf")
input_counts_deduplicated <- file.path(projectDir, "deduplicated_clearBox/readCounts/deduplicated_readCounts.txt")
input_metadata_file <- file.path(projectDir, "clearBoxRawData/cleanSampleMetadata.csv")

outdir <- file.path(projectDir, "deduplicated_clearBox/output")
output_all_cibersort_file <- file.path(outdir, "tpm_bulkCountMatrix_all_dedup.txt")
output_all_cibersort_file_gene_symbol <- file.path(outdir, "tpm_bulkCountMatrix_all_gene_symbol_dedup.txt")
output_all_cpm_file <- file.path(outdir, "cpm_dedup.txt")
output_all_log2cpm_file <- file.path(outdir, "log2cpm_dedup.txt")
output_deduplicated_log2cpms_no_gene_names_file <- file.path(outdir, "deduplicated_log2cpms_no_gene_names.txt")
output_dedup_no_split_ensembl_log2cpm_file <- file.path(outdir, "dedup_no_split_ensembl_log2cpm.txt")

dedup_samples <- c("STP0148", "SCP3954", "STP1206", "STP0932", "SCP4733", "STP0596")
dir.create(outdir)

# set max digits
options(digits=3)

# set filtering criteria
filterCPM <- 2
numSamples <- 5
```

Before I can transform my raw counts to TPMs I need to calculate the gene lengths from the original .gtf file.   

```{r}

# Try it
# Specifics of calculating rpkm: https://support.bioconductor.org/p/64585/

# 1. Get gene lengths
# https://www.biostars.org/p/83901/
# Or using GTFtools, but keep chromosome limitations in mind:
# http://genomespot.blogspot.com/2019/01/using-gtf-tools-to-get-gene-lengths.html
# GTFtools offers flexibility about how to calculate the gene length (median, mean, etc.)

# Note, this is the gtf used in the RNA-Seq pipeline to annotate hg38
txdb <- GenomicFeatures::makeTxDbFromGFF(file = file.path(genome_gtf_infile), format = "gtf")
# Collect exons per gene id
exon_list_per_gene <- GenomicFeatures::exonsBy(txdb, by = "gene")
exon_list_per_gene

```

```{r}

# For each gene, reduce all exons to a set of non-overlapping exons, calculate their length (widths) and sum them
exonic_gene_sizes <- 
  sum(IRanges::width(IRanges::reduce(exon_list_per_gene))) %>% 
  tibble::as_tibble(rownames = NA) %>% 
  tibble::rownames_to_column(var = "gene") %>%
  dplyr::select(., gene, length = value)
exonic_gene_sizes

# calculate the number of unique genes in the "gene" column of the exonic_gene_sizes data frame.
dim(exonic_gene_sizes) #58721 2
dim(exonic_gene_sizes[!duplicated(exonic_gene_sizes$gene), ]) #58721 2
# at this point there are no duplicated ensembl ids

```

## Import metadata
- import the metadata then drop samples not yet deduplicated

```{r}
## Metadata
metadata <- read_csv(file = input_metadata_file)
dedup_metadata <- dplyr::filter(metadata, samplename %in% dedup_samples)
```

## Import count files

```{r import gene counts - deduplicated}

# import the counts table
# counts are annotated to GRCh38
rawCounts_deduplicated <- read.delim(file = input_counts_deduplicated) %>%
  as.data.frame()
# tidy up the column names
colnames(rawCounts_deduplicated) <- gsub("X.scratch.user.smit1924.preeclampsia_sex_chromosome_informed.deduplicated_clearBox.deduplicated_bams.|_marked_duplicates.bam",
                                 "",
                                 colnames(rawCounts_deduplicated))
# remove the genes from the PAR_Y (all have zero counts) they look like "ENSG00000185203.12_PAR_Y"
rawCounts_deduplicated %<>% dplyr::filter(., !grepl("PAR_Y",Geneid))
dim(rawCounts_deduplicated) #58686 7
# calculate the number of unique ensembl ids
dim(rawCounts_deduplicated[!duplicated(rawCounts_deduplicated$Geneid), ]) #58676 7
# at this point there are no duplicated ensembl ids

ENSG00000186652_dedup <- rawCounts_deduplicated[str_detect(rawCounts_deduplicated$Geneid, "^ENSG00000186652"), ]
ENSG00000210082_dedup <- rawCounts_deduplicated[str_detect(rawCounts_deduplicated$Geneid, "^ENSG00000210082"), ]
ENSG00000183668_dedup <- rawCounts_deduplicated[str_detect(rawCounts_deduplicated$Geneid, "^ENSG00000183668"), ]
```

```{r}

# 2. Process raw count data in edgeR again, as per melanieRawCountsProcessing_v2.0.3.rmd
# Add gene length to expression data
raw_data <- rawCounts_deduplicated %>%
  dplyr::mutate(Length = exonic_gene_sizes$length[match(Geneid, exonic_gene_sizes$gene)])
# move the ensemble ids into the rownames and check table diminsions
raw_data <- raw_data %>% 
  tibble::column_to_rownames("Geneid")
head(raw_data)
dim(raw_data) # 58676 7

# establish the DGEList object
y <- edgeR::DGEList(counts = raw_data %>% dplyr::select(-Length),
                    genes = raw_data %>% dplyr::select(Length))

y
y$genes # A Length column is present 
dim(y) # 58676 6

head(y$counts) # Gene names are set as rownames
y$samples$lib.size # 11241651  5144473 14027324  5676726  6370904  5921367


# # Identify genes expressed at > 1 cpm in at least one sample
# keep <- rowSums(cpm(y) > filterCPM) >= numSamples
# table(keep) # Keeping 12367 genes
# 
# # Apply filter
# y <- y[keep, , keep.lib.sizes = FALSE]
# dim(y) # 12367    6
y$samples$lib.size #11089512  5095729 13877449  5615106  6319701  5878062

# TMM normalisation 
y$samples$norm.factors # currently all normalisation factors == 1
y <- calcNormFactors(y, method = "TMM")
y$samples$norm.factors # 1.104 0.915 1.080 1.123 0.869 0.940

# Now calculate RPKM: https://support.bioconductor.org/p/64585/
rpkm <- rpkm(y)

# And TPM: https://www.biostars.org/p/388584/#394579
tpm <- t( t(rpkm) / colSums(rpkm) ) * 1e6
# These TPM values now incorporate the TMM factors

```

## Generate and save a table of log2 cpms before adding the gene level information

```{r}
# generate the logged cpms
log2cpms_no_ensembl_split_dedup <- edgeR::cpm(y, log = TRUE)

# save a copy of the logged cpms
log2cpms_no_ensembl_split_dedup %>%
  as.data.frame() %>%
  # move the rownames into a column
  tibble::rownames_to_column("Geneid") %>%
  # save log2 cpm count file
  write.table(file = output_dedup_no_split_ensembl_log2cpm_file,
              quote = FALSE,
              sep = "\t",
              col.names = TRUE,
              row.names = FALSE)
```

## lets add the gene name information

### Import gencode V29 gff3 file

```{r}

# import the gencodev32_gff_inFile file
all_gencode_v29 <- rtracklayer::import(gencodev29_gff_inFile)
# this file contains more information than we need here
# subset out only the columns we need
gene_data <- data.frame(ensembl_gene_id = all_gencode_v29@elementMetadata$gene_id,
                        hgnc_symbol = all_gencode_v29@elementMetadata$gene_name,
                        seqnames = all_gencode_v29@seqnames)
# we're left with multiple identical rows, keep one each
gencode_v29_gene_id_symbol_chr_biotype <- gene_data %>%
  dplyr::distinct(, .keep_all = TRUE)
dim(gencode_v29_gene_id_symbol_chr_biotype) # 64837 3
dim(gencode_v29_gene_id_symbol_chr_biotype[!duplicated(gencode_v29_gene_id_symbol_chr_biotype$ensembl_gene_id), ]) # 64792 3

# remove duplicates
gencode_v29_gene_id_symbol_chr_biotype <- gencode_v29_gene_id_symbol_chr_biotype[!duplicated(gencode_v29_gene_id_symbol_chr_biotype$ensembl_gene_id), ]
```

## Generate and save a table of log2 cpms before adding the gene level information

```{r}
# generate the logged cpms
deduplicated_log2cpms_no_gene_names <- edgeR::cpm(y, log = TRUE)

# save a copy of the logged cpms
deduplicated_log2cpms_no_gene_names %>%
  as.data.frame() %>%
  # move the rownames into a column
  tibble::rownames_to_column("Geneid") %>%
  # save log2 cpm count file for CIBERSTORTx
  write.table(file = output_deduplicated_log2cpms_no_gene_names_file,
              quote = FALSE,
              sep = "\t",
              col.names = TRUE,
              row.names = FALSE)
```


### add the gene info to the tpms

```{r}
# join the counts with the gene info
tpm_gene_id_symbol_chr_biotype <- data.frame(tpm) %>%
  tibble::rownames_to_column("ensembl_gene_id") %>%
  dplyr::left_join(., gencode_v29_gene_id_symbol_chr_biotype, by = "ensembl_gene_id") %>%
  dplyr::select(colnames(gencode_v29_gene_id_symbol_chr_biotype), colnames(data.frame(tpm)))

dim(tpm_gene_id_symbol_chr_biotype)

# save a copy of the tpm counts with the gene info
tpm_gene_id_symbol_chr_biotype %>%
  # remove the number after the period
  tidyr::separate(., col = ensembl_gene_id, into = c("gene", "right")) %>% 
  dplyr::select(., -right) %>%
  # save TPM count file for CIBERSTORTx
  write.table(file = output_all_cibersort_file,
              quote = FALSE,
              sep = "\t",
              col.names = TRUE,
              row.names = FALSE)

# save a copy of the tpm counts with only the hgnc gene symbol
tpm_gene_id_symbol_chr_biotype %>%
  # remove the ensembl gene id and the chromosome name
  dplyr::select(., -ensembl_gene_id, -seqnames) %>%
  # save TPM count file for CIBERSTORTx
  write.table(file = output_all_cibersort_file_gene_symbol,
              quote = FALSE,
              sep = "\t",
              col.names = TRUE,
              row.names = FALSE)

```

## Generate CPM and log2CPM matrix and write to file

```{r}
# generate the cpms (not logged)
deduplicated_cpms <- edgeR::cpm(y, log = FALSE)

# join the cpms with the gene info
cpm_gene_id_symbol_chr <- data.frame(deduplicated_cpms) %>%
  tibble::rownames_to_column("ensembl_gene_id") %>%
  dplyr::left_join(., gencode_v29_gene_id_symbol_chr_biotype, by = "ensembl_gene_id") %>%
  dplyr::select(colnames(gencode_v29_gene_id_symbol_chr_biotype), colnames(data.frame(deduplicated_cpms)))

# save a copy of the cpm counts
cpm_gene_id_symbol_chr %>%
  as.data.frame() %>%
  # remove the number after the period
  tidyr::separate(., col = ensembl_gene_id, into = c("gene", "right")) %>% 
  dplyr::select(., -right) %>%
  # save TPM count file for CIBERSTORTx
  write.table(file = output_all_cpm_file,
              quote = FALSE,
              sep = "\t",
              col.names = TRUE,
              row.names = FALSE)

# generate the logged cpms
deduplicated_log2cpms <- edgeR::cpm(y, log = TRUE)
# join the log2cpms with the gene info
log2cpm_gene_id_symbol_chr <- data.frame(deduplicated_log2cpms) %>%
  tibble::rownames_to_column("ensembl_gene_id") %>%
  dplyr::left_join(., gencode_v29_gene_id_symbol_chr_biotype, by = "ensembl_gene_id") %>%
  dplyr::select(colnames(gencode_v29_gene_id_symbol_chr_biotype), colnames(data.frame(deduplicated_log2cpms)))


# save a copy of the logged cpms
log2cpm_gene_id_symbol_chr %>%
  # remove the number after the period
  tidyr::separate(., col = ensembl_gene_id, into = c("gene", "right")) %>% 
  dplyr::select(., -right) %>%
  # save TPM count file for CIBERSTORTx
  write.table(file = output_all_log2cpm_file,
              quote = FALSE,
              sep = "\t",
              col.names = TRUE,
              row.names = FALSE)
```