---
title: "campbell full matrix"
author: "Melanie Smith"
date: "1 September 2023"
output: html_document
---

```{r}
library(Seurat)
library(Matrix)
# library(tidyverse)
library(readxl)
library(readr)
library(magrittr)

library(ggplot2)
library(cowplot)
library(RColorBrewer)

# install.packages("remotes")
# remotes::install_github("MarianSchoen/DMC")
# library(DMC)

# set project directory
projectDir <- "/home/smit1924/preeclampsia_sex_chromosome_informed"
expt_name <- "campbell_full_refMatrix"

# infiles
campbell_inFile <- file.path(projectDir, "scRNAseqRawData/campbell/GSE182381_reference_sample.txt") 
gencodev32_inFile <- file.path(projectDir, "clearBoxRawData/gencode_v32_gene_id_vs_symbol.csv")

# outfiles
outdir <- file.path(projectDir, paste0(expt_name, "_output"))
output_cibersort_file <- file.path(outdir, paste0(expt_name, "new_referenceMatrix.txt"))

dir.create(outdir)

# set max digits
options(digits=3)
```

# Import data
```{r}
# import the martix
refMatrix <- read_delim(file = campbell_inFile,
                        col_names = FALSE,
                        delim = "\t")
```

# Generate some basic metadata
```{r}
metadata <- data.frame(orig.ident = sprintf("cell_%d", 1:length(colnames(refMatrix[, -1]))),
                       cell.type = t(refMatrix[1, -1]))
rownames(metadata) <- sprintf("cell_%d", 1:length(colnames(refMatrix[, -1])))
```

# Clean up the refMatrix
```{r}
refMatrix <- refMatrix %>%
  # cut the first row (contains the original file column header)
  dplyr::slice(-1) %>%
  # Put the gene names into the rownames
  tibble::column_to_rownames("X1")
# add column names to match the metadata
colnames(refMatrix) <- metadata$orig.ident
# make sure numeric
refMatrix <- refMatrix %>%
  dplyr::mutate_if(is.character, as.numeric)
# check class
# sapply(refMatrix, class)
dim(refMatrix)

# cut any rows that are all zeros
refMatrix <- refMatrix[rowSums(refMatrix[])>0,]
dim(refMatrix)

```

# Import gene information
```{r}
## Gene ID versus GeneSymbol
gencodev32 <- read_csv(file = gencodev32_inFile)
# df with matched gene symbols and Ensembl IDs
# geneNames <- data.frame(rownames(campbell_placenta10x)) %>% 
#   set_colnames("gene") %>% 
#   dplyr::left_join(., gencodev32, by = c("gene" = "hgnc_symbol"))
```

# Test if data is normalised
```{r}
# we don't know what has already been "done" to the data so lets see if it has been logged and normalised
# start with a subset of 400 cells
ncells <- 400
# set up colours for the density plots
col_cells <- colorRampPalette(brewer.pal(11,"Spectral"))(100)
# png(filename=file.path(outdir, "un_normalised.png"),
#     width=12,
#     height=7,
#     units = "in",
#     res = 150)
boxplot(refMatrix[, 1:ncells], las = 2,
        col = col_cells,
        main = "")
title(main = "Full Campbell ref matrix(1:400 cells): Un-normalised data ",
      ylab = "counts") 
# dev.off()

# okay - data is not logged and normalised - next step is to create the seurat object and perform some basic QC and normalisation
```

# Create minimum seurat object
```{r}
# create a new seurat object using the count table and self-generated metadata
campbell <- CreateSeuratObject(refMatrix, meta.data = metadata)
```

# Pre-process
```{r}
# quantify mitochondrial gene content (MT-) genes
campbell[["percent.mt"]] <- PercentageFeatureSet(campbell, pattern = "^MT-")

# visualise QC metric as a violin plot
VlnPlot(campbell, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)

plot1 <- FeatureScatter(campbell, feature1 = "nCount_RNA", feature2 = "percent.mt")
plot2 <- FeatureScatter(campbell, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")

# print the QC plots
plot1 + plot2

# Subset out the poor quality cells
campbell <- subset(campbell, subset = nFeature_RNA > 500 &
                        nFeature_RNA < 2500 & percent.mt < 5)
# Normalise the data
campbell <- NormalizeData(object = campbell,
                          normalization.method = "RC",
                          scale.factor = 1e6,
                             verbose = FALSE)

# Identify highly variable features
campbell <- FindVariableFeatures(object = campbell,
                                      nfeatures = 6000,
                                      verbose = FALSE,
                                      selection.method = 'vst')
# plot the highly variable features
top10 <- head(VariableFeatures(campbell), 10)
plot3 <- VariableFeaturePlot(campbell)

plot4 <- LabelPoints(plot = plot3, points = top10, repel = TRUE)
plot3 + plot4

all.genes <- rownames(campbell)
campbell <- ScaleData(campbell,
                           features = all.genes,
                           verbose = FALSE)
```

## Plot normalised data
```{r}
# png(filename=file.path(outdir, "un_normalised.png"),
#     width=12,
#     height=7,
#     units = "in",
#     res = 150)
boxplot(refMatrix[, 1:ncells], las = 2,
        col = col_cells,
        main = "")
title(main = "Full Campbell ref matrix(1:400 cells): Un-normalised data ",
      ylab = "counts") 
# dev.off()
```

```{r}

campbell <- RunPCA(campbell,
                   npcs = 20,
                   verbose = FALSE)
campbell <- RunUMAP(campbell,
                    reduction = "pca",
                    dims = 1:20)
# visualise the dimensional loadings
VizDimLoadings(campbell, dims = 1:2, reduction = "pca")
# plot the PCA
DimPlot(campbell, reduction = "pca")
# visualise the first 4 PCs as heatmap of 500 "most extreme" genes
DimHeatmap(campbell, dims = 1:4, cells = 500, balanced = TRUE)

# Determine the dimensionality of the dataset
campbell <- JackStraw(campbell, num.replicate = 100)
campbell <- ScoreJackStraw(campbell, dims = 1:20)

# visualise
JackStrawPlot(campbell, dims = 1:20)

# saveRDS(campbell, file = file.path(outdir, "full_seurat_campbell.rds"))
# 
# campbell <- readRDS(file = file.path(outdir, "full_seurat_campbell.rds"))
  
  
```

# Plot umap
```{r}
png(filename=file.path(outdir, "refMatrix_umap.png"),
    width=12,
    height=7,
    units = "in",
    res = 150)
DimPlot(campbell, group.by = "cell.type", reduction = "umap", label = TRUE) + NoLegend()
dev.off()
```

# Create signature Matrix
```{r}
# # this function is from here https://rdrr.io/github/MarianSchoen/DMC/f/README.md
# create_sig_matrix <- function(
#   exprs,
#   pheno,
#   exclude.celltypes = NULL,
#   max.genes = NULL,
#   cell.type.column = "cell_type"
#   ) {
#   # parameter checks
#   if (nrow(pheno) != ncol(exprs)) {
#       stop("Number of columns in exprs and rows in pheno do not match")
#   }
#   if (!is.null(max.genes) && max.genes == 0) {
#         max.genes <- NULL
#   }
#   rownames(pheno) <- colnames(exprs)
# 
#   # exclude specified cell types
#   if (!is.null(exclude.celltypes)) {
#     to.exclude <- which(pheno[, cell.type.column] %in% exclude.celltypes)
#     if (length(to.exclude) > 0) {
#       exprs <- exprs[, -to.exclude, drop = F]
#       pheno <- pheno[-to.exclude, , drop = F]
#     }
#   }
# 
#   # make sure that not more genes than available are selected
#   if (is.null(max.genes)) {
# 	   max.genes <- nrow(exprs)
#   }
#   #cat("Maximum number of genes per cell type: ", max.genes, "\n")
# 
#   # for each cell type test against all others for DEG
#   # using two-sided t-test
#   deg.per.type <- list()
#   for (t in unique(pheno[, cell.type.column])) {
#     labs <- ifelse(pheno[, cell.type.column] == t, 0, 1)
#     t.test.result <- multtest::mt.teststat(
#       X = exprs,
#       classlabel = labs,
#       test = "t"
#     )
#     p.vals <- 2 * pt(abs(t.test.result), length(labs) - 2, lower.tail = FALSE)
#     names(p.vals) <- rownames(exprs)
# 
#     nna.genes <- names(p.vals)[!is.na(p.vals)]
#     p.vals <- p.vals[nna.genes]
# 
#     # control FDR either via q-value or with benjamini-hochberg
#     q.vals <- try({
#       qvalue::qvalue(p.vals)$qvalues
#     }, silent = TRUE)
#     if (class(q.vals) == "try-error") {
#     	#warning("q-value calculation failed. Using BH-correction.")
#     	q.vals <- p.adjust(p.vals, "BH")
#       sig.entries <- which(q.vals < 0.1)
#     }else{
#       sig.entries <- which(q.vals < 0.3)
#     }
#     sig.genes <- nna.genes[sig.entries]
# 
#     # catch possible errors related to sig.genes
#     if (any(is.na(sig.genes))) {
#       sig.entries <- sig.entries[!is.na(sig.genes)]
#       sig.genes <- sig.genes[!is.na(sig.genes)]
#     }
#     if (!length(sig.genes) > 0) break
# 
#     # calculate fold changes
#     fold.changes <- log2(Matrix::rowMeans(
#         exprs[sig.entries, which(labs == 0), drop = F]
#       )) - log2(Matrix::rowMeans(
#         exprs[sig.entries, which(labs == 1), drop = F]
#       ))
#     if (any(is.infinite(fold.changes))) {
#       sig.genes <- sig.genes[-which(is.infinite(fold.changes))]
#       fold.changes <- fold.changes[-which(is.infinite(fold.changes))]
#     }
#     if (any(is.nan(fold.changes))) {
#       sig.genes <- sig.genes[-which(is.nan(fold.changes))]
#       fold.changes <- fold.changes[-which(is.nan(fold.changes))]
#     }
#     # add genes for each type ordered by decreasing fold change
#     deg.per.type[[t]] <- sig.genes[order(abs(fold.changes), decreasing = TRUE)]
#   }
# 
#   # reduce to one reference profile per cell type
#   ref.profiles <- matrix(
#     0,
#     nrow = nrow(exprs),
#     ncol = length(unique(pheno[, cell.type.column]))
#   )
#   colnames(ref.profiles) <- unique(pheno[, cell.type.column])
#   rownames(ref.profiles) <- rownames(exprs)
# 
#   for (t in colnames(ref.profiles)) {
#     ref.profiles[, t] <- Matrix::rowMeans(
#       exprs[, which(pheno[, cell.type.column] == t), drop = F]
#     )
#   }
# 
#   # again following Newman et al.:
#   # take top g genes for every cell type, create signature matrices
#   # choose the gene set that minimizes condition number
#   if (length(deg.per.type) > 0) {
#   limit <- min(
#     max(sapply(deg.per.type, length)), max.genes
#   )
#   }else{
#     warning("No significant genes found. Returning NULL.")
#     return(NULL)
#   }
# 
#   cond.nums <- rep(Inf, times = limit)
#   # condition number is unstable for very few genes, therefore use at least 10
#   for (g in 10:limit) {
#     all.genes <- unique(unlist(
#       sapply(deg.per.type, function(sub.genes, lim = g) {
#         sub.genes[1:min(length(sub.genes), lim)]
#       })
#     ))
#     if (any(is.na(all.genes))) {
#       all.genes <- all.genes[-which(is.na(all.genes))]
#     }
# 
#     # estimate condition number of reduced matrix
#     cond.nums[g] <- kappa(ref.profiles[all.genes,, drop = FALSE], exact = FALSE)
#   }
#   optimal.g <- which.min(cond.nums)
# 
#   # alternative:
#   #optimal.g <- limit
#   #cat(
#   #  "Chose ", optimal.g,
#   #  "genes per cell type resulting in condition number of",
#   #  cond.nums[optimal.g], "\n"
#   #)
# 
#   # create gene list with the optimal g
#   optimal.genes <- unique(unlist(
#     sapply(deg.per.type, function(sub.genes, opt.g = optimal.g) {
#       sub.genes[1:min(opt.g, length(sub.genes))]
#     })
#   ))
# 
#   ref.profiles <- ref.profiles[optimal.genes, ]
# 
#   # remove any duplicate genes that might be in the matrix
#   if (any(duplicated(rownames(ref.profiles)))) {
#     warning("Found duplicates in reference matrix")
#     ref.profiles <- ref.profiles[-which(duplicated(rownames(ref.profiles))), ]
#   }
#   return(ref.profiles)
# }
# 
# # need to make the matrix of counts smaller
# # what are the most variable genes
# var_genes <- VariableFeatures(campbell)
# # subset counts to only contain most variable gens
# var_campbell <- GetAssayData(campbell)[var_genes,]
# # pull the metadata table
# campbell_meta <- campbell@meta.data
# # garbage collection
# gc()
# 
# # create signature matrix
# signatureMatrix <- create_sig_matrix(
#   var_campbell,
#   campbell_meta,
#   exclude.celltypes = NULL,
#   max.genes = NULL,
#   cell.type.column = "cell.type"
# )
# 
# geneNames <- signatureMatrix %>% 
#   data.frame() %>% 
#   tibble::rownames_to_column("gene") %>%
#   dplyr::select(gene) %>%
#   dplyr::left_join(., gencodev32, by = c("gene" = "hgnc_symbol"))
# 
# testMatrix <- data.frame(signatureMatrix) %>%
#   tibble::rownames_to_column("gene") %>%
#   dplyr::left_join(., geneNames, by = "gene") %>%
#   dplyr::select(., -gene) %>%
#   dplyr::select(., GeneSymbol = ensembl, everything())
# 
# # write out signature matrix
# # save CPM count file for CIBERSTORTx
# testMatrix %>%
#   write.table(file = output_cibersort_file,
#               quote = FALSE,
#               sep = "\t",
#               col.names = TRUE,
#               row.names = FALSE)
```

# Create reference matrix
```{r}
campbell <- NormalizeData(campbell,
                          normalization.method = "RC",
                          scale.factor = 1e6)
# find variable features
campbell <- FindVariableFeatures(campbell,
                                 selection.method = "vst",
                                 nfeatures = 6000)
```







